**Relational Tables and Joining Them**

Activating SQL magic extension.

In [1]:
%load_ext sql

# INNER JOIN
Combining results where the **ID** clause is present in both tables.

<img src="images/inner_join.png" alt='Drawing' style="width: 200px;"/>

In [3]:
%sql postgresql://postgres:taskuarvuti@localhost:5432/postgres

'Connected: postgres@postgres'

## ON

Combine *cities* and *countries* tables **on** *country_code* & *code*.

In [5]:
%%sql
select *
from cities
inner join countries 
on cities.country_code = countries.code
limit 3

 * postgresql://postgres:***@localhost:5432/postgres
3 rows affected.


[('Abidjan', 'CIV', 4765000.0, None, 4765000.0, 'CIV', "Cote d'Ivoire", 'Africa', 'Western Africa', 322463.0, 1960, 'Cote d\x92Ivoire', 'Republic', 'Yamoussoukro', -4.0305, 5.332),
 ('Abu Dhabi', 'ARE', 1145000.0, None, 1145000.0, 'ARE', 'United Arab Emirates', 'Asia', 'Middle East', 83600.0, 1971, 'Al-Imarat al-´Arabiya al-Muttahida', 'Emirate Federation', 'Abu Dhabi', 54.3705, 24.4764),
 ('Abuja', 'NGA', 1235880.0, 6000000.0, 1235880.0, 'NGA', 'Nigeria', 'Africa', 'Western Africa', 923768.0, 1960, 'Nigeria', 'Federal Republic', 'Abuja', 7.48906, 9.05804)]

Let's try to find information about Estonia.

In [8]:
%%sql
select *
from cities
inner join countries
on cities.country_code = countries.code
where cities.country_code = 'EST'

 * postgresql://postgres:***@localhost:5432/postgres
0 rows affected.


[]

In [9]:
%%sql
select * from countries where code='EST'

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[('EST', 'Estonia', 'Europe', 'Baltic Countries', 45227.0, 1991, 'Eesti', 'Republic', 'Tallinn', 24.7586, 59.4392)]

Return name of the city, country, region.

In [10]:
%%sql
select cities.name, countries.name, countries.region
from cities
inner join countries
on cities.country_code = countries.code
limit 4

 * postgresql://postgres:***@localhost:5432/postgres
4 rows affected.


[('Abidjan', "Cote d'Ivoire", 'Western Africa'),
 ('Abu Dhabi', 'United Arab Emirates', 'Middle East'),
 ('Abuja', 'Nigeria', 'Western Africa'),
 ('Accra', 'Ghana', 'Western Africa')]

## Table aliasing

In [11]:
%%sql
select ci.name, co.name, co.region
from cities as ci
inner join countries as co
on ci.country_code = co.code
limit 4

 * postgresql://postgres:***@localhost:5432/postgres
4 rows affected.


[('Abidjan', "Cote d'Ivoire", 'Western Africa'),
 ('Abu Dhabi', 'United Arab Emirates', 'Middle East'),
 ('Abuja', 'Nigeria', 'Western Africa'),
 ('Accra', 'Ghana', 'Western Africa')]

Return inflation rate for 2010 and 2015 from *countries* and *economies* tables.

In [22]:
%%sql
select co.name, ec.year, ec.inflation_rate
from countries as co
inner join economies as ec
on co.code = ec.code
limit 4

 * postgresql://postgres:***@localhost:5432/postgres
4 rows affected.


[('Afghanistan', 2010, 2.179),
 ('Afghanistan', 2015, -1.549),
 ('Angola', 2010, 14.48),
 ('Angola', 2015, 10.287)]

## Multiple joins

Return for each country the country name, its region, and the fertility rate and unemployment rate for both 2010 and 2015.

In [26]:
%%sql
select co.name, co.region, po.year, po.fertility_rate, ec.unemployment_rate
from countries as co
inner join populations as po
on co.code=po.country_code
inner join economies as ec
on co.code=ec.code and po.year=ec.year
limit 7

 * postgresql://postgres:***@localhost:5432/postgres
7 rows affected.


[('Afghanistan', 'Southern and Central Asia', 2010, 5.746, None),
 ('Afghanistan', 'Southern and Central Asia', 2015, 4.653, None),
 ('Angola', 'Central Africa', 2010, 6.416, None),
 ('Angola', 'Central Africa', 2015, 5.996, None),
 ('Albania', 'Southern Europe', 2010, 1.663, 14.0),
 ('Albania', 'Southern Europe', 2015, 1.793, 17.1),
 ('United Arab Emirates', 'Middle East', 2010, 1.868, None)]

If we do not specify <code>and po.year=ec.year</code> we duplicate the *year* field from populations and economies:

In [27]:
%%sql
select co.name, region, po.year, fertility_rate, ec.unemployment_rate
from countries as co
    inner join populations as po
        on co.code=po.country_code
    inner join economies as ec
        on co.code=ec.code
limit 7

 * postgresql://postgres:***@localhost:5432/postgres
7 rows affected.


[('Afghanistan', 'Southern and Central Asia', 2015, 4.653, None),
 ('Afghanistan', 'Southern and Central Asia', 2010, 5.746, None),
 ('Afghanistan', 'Southern and Central Asia', 2015, 4.653, None),
 ('Afghanistan', 'Southern and Central Asia', 2010, 5.746, None),
 ('Angola', 'Central Africa', 2015, 5.996, None),
 ('Angola', 'Central Africa', 2010, 6.416, None),
 ('Angola', 'Central Africa', 2015, 5.996, None)]

## USING()
If we have the exact same **ID** field name in both or multiple tables we can use funtion **USING()** for specifiyng the ID field name.

In [29]:
%%sql
--Countries name as country, continent, language as language and if language is official or not.
select co.name as country, continent, la.name as language, official
from countries as co
    inner join languages as la
        USING(code)
limit 10

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Afghanistan', 'Asia', 'Dari', True),
 ('Afghanistan', 'Asia', 'Pashto', True),
 ('Afghanistan', 'Asia', 'Turkic', False),
 ('Afghanistan', 'Asia', 'Other', False),
 ('Albania', 'Europe', 'Albanian', True),
 ('Albania', 'Europe', 'Greek', False),
 ('Albania', 'Europe', 'Other', False),
 ('Albania', 'Europe', 'unspecified', False),
 ('Algeria', 'Africa', 'Arabic', True),
 ('Algeria', 'Africa', 'French', False)]

## Self-join
Compare fields in the same table.

In [62]:
%%sql
--Calculate the percentage increase in population 
--from 2010 to 2015 for each country code!

select po1.country_code, 
       po1.size as pop_2010, 
       po2.size as pop_2015, 
       ROUND(((po2.size - po1.size)/po1.size * 100.0)) as pop_pct_inrcease
from populations as po1
    inner join populations as po2
        on po1.country_code=po2.country_code
            and po1.year=2010 and po2.year=2015
limit 5

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('ABW', 101597.0, 103889.0, 2.0),
 ('AFG', 27962200.0, 32526600.0, 16.0),
 ('AGO', 21220000.0, 25022000.0, 18.0),
 ('ALB', 2913020.0, 2889170.0, -1.0),
 ('AND', 84419.0, 70473.0, -17.0)]

## CASE WHEN
For categorising data we can use following <code>SQL</code> syntax:<br>
* <code>WHEN</code>
* <code>THEN</code>
* <code>ELSE</code>
* <code>END</code>

Using the <code>countries</code> table, create a new field **geosize_group** that groups the countries into three groups:<br>
* <code>surface_area</code> > 2m = 'large'
* 2m > <code>surface_area</code> > 350k = 'medium'
* otherwise <code>surface_area</code> = 'small'

In [72]:
%%sql
select name, surface_area,
    CASE WHEN surface_area > 2e6 THEN 'large'
        WHEN surface_area > 3.5e5 and surface_area < 2e6 THEN 'medium'
        ELSE 'small' END
        as geosize_group
from countries
limit 4

 * postgresql://postgres:***@localhost:5432/postgres
4 rows affected.


[('Afghanistan', 652090.0, 'medium'),
 ('Netherlands', 41526.0, 'small'),
 ('Albania', 28748.0, 'small'),
 ('Algeria', 2381740.0, 'large')]

## INTO
Clause into returns a query into a seprate table in the DB.

In [81]:
%%sql
select name, surface_area,
    CASE WHEN surface_area > 2e6 THEN 'large'
        WHEN surface_area > 3.5e5 THEN 'medium'
        ELSE 'small' END
        as geosize_group
into geogropued_countries
from countries

 * postgresql://postgres:***@localhost:5432/postgres
206 rows affected.


[]

**Assining a query to a variable but NOT creating a new table**<br>
Jupyter <code>sql magic</code> way:

In [76]:
%%sql geo_countries <<
select name, surface_area,
    CASE WHEN surface_area > 2e6 THEN 'large'
        WHEN surface_area > 3.5e5 THEN 'medium'
        ELSE 'small' END
        as geosize_group
from countries

 * postgresql://postgres:***@localhost:5432/postgres
206 rows affected.
Returning data to local variable geo_countries


Using the populations table focused only for the year 2015, create a new field AS popsize_group to organize population size into

*    'large' (> 50 million),
*    'medium' (> 1 million), and
*    'small' groups.

Select only the country code, population size, and this new popsize_group as fields.

In [91]:
%%sql
select country_code,size,
    CASE WHEN size >= 5e7 THEN 'large'
        WHEN size >= 1e6 THEN 'medium'
        ELSE 'small' END
        as popsize_group
into pop_plus
from populations
where year = 2015;

 * postgresql://postgres:***@localhost:5432/postgres
217 rows affected.


[]

In [92]:
%%sql
select *
from pop_plus
limit 7

 * postgresql://postgres:***@localhost:5432/postgres
7 rows affected.


[('ABW', 103889.0, 'small'),
 ('AFG', 32526600.0, 'medium'),
 ('AGO', 25022000.0, 'medium'),
 ('ALB', 2889170.0, 'medium'),
 ('AND', 70473.0, 'small'),
 ('ARE', 9156960.0, 'medium'),
 ('ARG', 43416800.0, 'medium')]

* Keep the first query intact that creates pop_plus using INTO. 
* Write a query to join countries_plus AS c on the left with pop_plus AS p on the right matching on the country code fields.
* Sort the data based on geosize_group, in ascending order so that large appears on top.
* Select the name, continent, geosize_group, and popsize_group fields.

In [95]:
%%sql
select name, continent, geosize_group, popsize_group
from countries_plus as c_plus
    inner join pop_plus as p_plus
        on c_plus.code=p_plus.country_code
order by geosize_group
limit 10

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Greenland', 'North America', 'large', 'small'),
 ('India', 'Asia', 'large', 'large'),
 ('Congo, The Democratic Republic of the', 'Africa', 'large', 'large'),
 ('Algeria', 'Africa', 'large', 'medium'),
 ('Canada', 'North America', 'large', 'medium'),
 ('Argentina', 'South America', 'large', 'medium'),
 ('China', 'Asia', 'large', 'large'),
 ('Australia', 'Oceania', 'large', 'medium'),
 ('Brazil', 'South America', 'large', 'large'),
 ('Kazakhstan', 'Asia', 'large', 'medium')]

# Outer joins
Keeping all of the records of the original table but leaving out the records of the joining table which **ID** value didn't match.
## LEFT JOIN
<img src="images/left_join.png" alt='Drawing' style="width: 200px;"/>

Let's compare <code>inner join</code> and <code>left join</code>

In [104]:
%%sql
--inner join
select ci.name as city, code,
       co.name as country, region,
       city_proper_pop
from cities as ci
    inner join countries as co
        on ci.country_code=co.code
order by code desc

 * postgresql://postgres:***@localhost:5432/postgres
230 rows affected.


[('Harare', 'ZWE', 'Zimbabwe', 'Eastern Africa', 1606000.0),
 ('Lusaka', 'ZMB', 'Zambia', 'Eastern Africa', 1742980.0),
 ('Cape Town', 'ZAF', 'South Africa', 'Southern Africa', 3740030.0),
 ('Johannesburg', 'ZAF', 'South Africa', 'Southern Africa', 4434830.0),
 ('Durban', 'ZAF', 'South Africa', 'Southern Africa', 3442360.0),
 ('Ekurhuleni', 'ZAF', 'South Africa', 'Southern Africa', 3178470.0),
 ("Sana'a", 'YEM', 'Yemen', 'Middle East', 1937450.0),
 ('Ho Chi Minh City', 'VNM', 'Vietnam', 'Southeast Asia', 7681700.0),
 ('Hanoi', 'VNM', 'Vietnam', 'Southeast Asia', 6844100.0),
 ('Caracas', 'VEN', 'Venezuela', 'South America', 1943900.0),
 ('Maracaibo', 'VEN', 'Venezuela', 'South America', 1599940.0),
 ('Tashkent', 'UZB', 'Uzbekistan', 'Southern and Central Asia', 2309600.0),
 ('Chicago', 'USA', 'United States', 'North America', 2695600.0),
 ('Los Angeles', 'USA', 'United States', 'North America', 3884310.0),
 ('Philadelphia', 'USA', 'United States', 'North America', 1567870.0),
 ('Phoenix', 'USA', 'United States', 'North America', 1563020.0),
 ('Dallas', 'USA', 'United States', 'North America', 1317930.0),
 ('New York City', 'USA', 'United States', 'North America', 8550400.0),
 ('Houston', 'USA', 'United States', 'North America', 2489560.0),
 ('San Diego', 'USA', 'United States', 'North America', 1394930.0),
 ('San Antonio', 'USA', 'United States', 'North America', 1469840.0),
 ('Montevideo', 'URY', 'Uruguay', 'South America', 1305080.0),
 ('Kiev', 'UKR', 'Ukraine', 'Eastern Europe', 2908700.0),
 ('Kharkov', 'UKR', 'Ukraine', 'Eastern Europe', 1439570.0),
 ('Kampala', 'UGA', 'Uganda', 'Eastern Africa', 1507080.0),
 ('Dar es Salaam', 'TZA', 'Tanzania', 'Eastern Africa', 4364540.0),
 ('Ankara', 'TUR', 'Turkey', 'Middle East', 5271000.0),
 ('Izmir', 'TUR', 'Turkey', 'Middle East', 4168000.0),
 ('Istanbul', 'TUR', 'Turkey', 'Middle East', 14025000.0),
 ('Tunis', 'TUN', 'Tunisia', 'Northern Africa', 1056250.0),
 ('Bangkok', 'THA', 'Thailand', 'Southeast Asia', 8280920.0),
 ('Singapore', 'SGP', 'Singapore', 'Southeast Asia', 5535000.0),
 ('Dakar', 'SEN', 'Senegal', 'Western Africa', 1146050.0),
 ('Khartoum', 'SDN', 'Sudan', 'Northern Africa', 3639600.0),
 ('Riyadh', 'SAU', 'Saudi Arabia', 'Middle East', 5676620.0),
 ('Jeddah', 'SAU', 'Saudi Arabia', 'Middle East', 3456260.0),
 ('Yekaterinburg', 'RUS', 'Russian Federation', 'Eastern Europe', 1428040.0),
 ('Nizhny Novgorod', 'RUS', 'Russian Federation', 'Eastern Europe', 1250620.0),
 ('Moscow', 'RUS', 'Russian Federation', 'Eastern Europe', 12197600.0),
 ('Rostov-on-Don', 'RUS', 'Russian Federation', 'Eastern Europe', 1119900.0),
 ('Omsk', 'RUS', 'Russian Federation', 'Eastern Europe', 1154120.0),
 ('Novosibirsk', 'RUS', 'Russian Federation', 'Eastern Europe', 1567090.0),
 ('Saint Petersburg', 'RUS', 'Russian Federation', 'Eastern Europe', 5191690.0),
 ('Pyongyang', 'PRK', 'North Korea', 'Eastern Asia', 3255390.0),
 ('Warsaw', 'POL', 'Poland', 'Eastern Europe', 1753980.0),
 ('Quezon City', 'PHL', 'Philippines', 'Southeast Asia', 2936120.0),
 ('Manila', 'PHL', 'Philippines', 'Southeast Asia', 1780150.0),
 ('Caloocan', 'PHL', 'Philippines', 'Southeast Asia', 1583980.0),
 ('Davao City', 'PHL', 'Philippines', 'Southeast Asia', 1632990.0),
 ('Lima', 'PER', 'Peru', 'South America', 8852000.0),
 ('Multan', 'PAK', 'Pakistan', 'Southern and Central Asia', 3117000.0),
 ('Hyderabad', 'PAK', 'Pakistan', 'Southern and Central Asia', 3429470.0),
 ('Islamabad', 'PAK', 'Pakistan', 'Southern and Central Asia', 1900000.0),
 ('Rawalpindi', 'PAK', 'Pakistan', 'Southern and Central Asia', 3198910.0),
 ('Karachi', 'PAK', 'Pakistan', 'Southern and Central Asia', 27506000.0),
 ('Gujranwala', 'PAK', 'Pakistan', 'Southern and Central Asia', 2700000.0),
 ('Lahore', 'PAK', 'Pakistan', 'Southern and Central Asia', 10355000.0),
 ('Faisalabad', 'PAK', 'Pakistan', 'Southern and Central Asia', 6480760.0),
 ('Peshawar', 'PAK', 'Pakistan', 'Southern and Central Asia', 3201000.0),
 ('Auckland', 'NZL', 'New Zealand', 'Austr

In [105]:
%%sql
select ci.name as city, code,
       co.name as country, region,
       city_proper_pop
from cities as ci
    left join countries as co
        on ci.country_code=co.code
order by code desc

 * postgresql://postgres:***@localhost:5432/postgres
236 rows affected.


[('Taichung', None, None, None, 2752410.0),
 ('Tainan', None, None, None, 1885250.0),
 ('Kaohsiung', None, None, None, 2778920.0),
 ('Bucharest', None, None, None, 1883420.0),
 ('Taipei', None, None, None, 2704970.0),
 ('New Taipei City', None, None, None, 3954930.0),
 ('Harare', 'ZWE', 'Zimbabwe', 'Eastern Africa', 1606000.0),
 ('Lusaka', 'ZMB', 'Zambia', 'Eastern Africa', 1742980.0),
 ('Cape Town', 'ZAF', 'South Africa', 'Southern Africa', 3740030.0),
 ('Ekurhuleni', 'ZAF', 'South Africa', 'Southern Africa', 3178470.0),
 ('Durban', 'ZAF', 'South Africa', 'Southern Africa', 3442360.0),
 ('Johannesburg', 'ZAF', 'South Africa', 'Southern Africa', 4434830.0),
 ("Sana'a", 'YEM', 'Yemen', 'Middle East', 1937450.0),
 ('Hanoi', 'VNM', 'Vietnam', 'Southeast Asia', 6844100.0),
 ('Ho Chi Minh City', 'VNM', 'Vietnam', 'Southeast Asia', 7681700.0),
 ('Maracaibo', 'VEN', 'Venezuela', 'South America', 1599940.0),
 ('Caracas', 'VEN', 'Venezuela', 'South America', 1943900.0),
 ('Tashkent', 'UZB', 'Uzbekistan', 'Southern and Central Asia', 2309600.0),
 ('Dallas', 'USA', 'United States', 'North America', 1317930.0),
 ('San Diego', 'USA', 'United States', 'North America', 1394930.0),
 ('New York City', 'USA', 'United States', 'North America', 8550400.0),
 ('Philadelphia', 'USA', 'United States', 'North America', 1567870.0),
 ('Los Angeles', 'USA', 'United States', 'North America', 3884310.0),
 ('Houston', 'USA', 'United States', 'North America', 2489560.0),
 ('Phoenix', 'USA', 'United States', 'North America', 1563020.0),
 ('San Antonio', 'USA', 'United States', 'North America', 1469840.0),
 ('Chicago', 'USA', 'United States', 'North America', 2695600.0),
 ('Montevideo', 'URY', 'Uruguay', 'South America', 1305080.0),
 ('Kharkov', 'UKR', 'Ukraine', 'Eastern Europe', 1439570.0),
 ('Kiev', 'UKR', 'Ukraine', 'Eastern Europe', 2908700.0),
 ('Kampala', 'UGA', 'Uganda', 'Eastern Africa', 1507080.0),
 ('Dar es Salaam', 'TZA', 'Tanzania', 'Eastern Africa', 4364540.0),
 ('Istanbul', 'TUR', 'Turkey', 'Middle East', 14025000.0),
 ('Izmir', 'TUR', 'Turkey', 'Middle East', 4168000.0),
 ('Ankara', 'TUR', 'Turkey', 'Middle East', 5271000.0),
 ('Tunis', 'TUN', 'Tunisia', 'Northern Africa', 1056250.0),
 ('Bangkok', 'THA', 'Thailand', 'Southeast Asia', 8280920.0),
 ('Singapore', 'SGP', 'Singapore', 'Southeast Asia', 5535000.0),
 ('Dakar', 'SEN', 'Senegal', 'Western Africa', 1146050.0),
 ('Khartoum', 'SDN', 'Sudan', 'Northern Africa', 3639600.0),
 ('Riyadh', 'SAU', 'Saudi Arabia', 'Middle East', 5676620.0),
 ('Jeddah', 'SAU', 'Saudi Arabia', 'Middle East', 3456260.0),
 ('Novosibirsk', 'RUS', 'Russian Federation', 'Eastern Europe', 1567090.0),
 ('Yekaterinburg', 'RUS', 'Russian Federation', 'Eastern Europe', 1428040.0),
 ('Moscow', 'RUS', 'Russian Federation', 'Eastern Europe', 12197600.0),
 ('Saint Petersburg', 'RUS', 'Russian Federation', 'Eastern Europe', 5191690.0),
 ('Rostov-on-Don', 'RUS', 'Russian Federation', 'Eastern Europe', 1119900.0),
 ('Omsk', 'RUS', 'Russian Federation', 'Eastern Europe', 1154120.0),
 ('Nizhny Novgorod', 'RUS', 'Russian Federation', 'Eastern Europe', 1250620.0),
 ('Pyongyang', 'PRK', 'North Korea', 'Eastern Asia', 3255390.0),
 ('Warsaw', 'POL', 'Poland', 'Eastern Europe', 1753980.0),
 ('Quezon City', 'PHL', 'Philippines', 'Southeast Asia', 2936120.0),
 ('Davao City', 'PHL', 'Philippines', 'Southeast Asia', 1632990.0),
 ('Caloocan', 'PHL', 'Philippines', 'Southeast Asia', 1583980.0),
 ('Manila', 'PHL', 'Philippines', 'Southeast Asia', 1780150.0),
 ('Lima', 'PER', 'Peru', 'South America', 8852000.0),
 ('Multan', 'PAK', 'Pakistan', 'Southern and Central Asia', 3117000.0),
 ('Gujranwala', 'PAK', 'Pakistan', 'Southern and Central Asia', 2700000.0),
 ('Peshawar', 'PAK', 'Pakistan', 'Southern and Central Asia', 3201000.0),
 ('Islamabad', 'PAK', 'Pakistan', 'Southern and Central Asia', 1900000.0),
 ('Hyderabad', 'PAK', 'Pakistan', 'Southern and Central Asia', 3429470.0),
 ('Karachi', 'PAK', 'Pakistan', 'Southern and Central Asia', 27506000.0)

Resulting **230** vs **236** records for <code>inner join</code> and <code>left join</code> respectively.

**Compare <code>inner join</code> and <code>left join</code> yet again**
1. Count the records in <code>countries</code> and <code>languages</code> tables.
2. Perform an <code>inner join</code>. Alias the name of the country field as *country* and the name of the language field as *language*. Sort based on descending country name.
3. Perform the same query with <code>left join</code>

In [112]:
%%sql
select
    (select count(*) as countries_records
     from countries),
    (select count(*) as languages_records
     from languages)

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[(206, 955)]

In [114]:
%%sql
select co.name as country, local_name,
       la.name as language, percent
from countries as co
    inner join languages as la
        using(code)
order by country desc

 * postgresql://postgres:***@localhost:5432/postgres
914 rows affected.


[('Zimbabwe', 'Zimbabwe', 'Shona', None),
 ('Zimbabwe', 'Zimbabwe', 'Tonga', None),
 ('Zimbabwe', 'Zimbabwe', 'Tswana', None),
 ('Zimbabwe', 'Zimbabwe', 'Venda', None),
 ('Zimbabwe', 'Zimbabwe', 'Xhosa', None),
 ('Zimbabwe', 'Zimbabwe', 'Sotho', None),
 ('Zimbabwe', 'Zimbabwe', 'sign', None),
 ('Zimbabwe', 'Zimbabwe', 'Shangani', None),
 ('Zimbabwe', 'Zimbabwe', 'Ndau', None),
 ('Zimbabwe', 'Zimbabwe', 'Nambya', None),
 ('Zimbabwe', 'Zimbabwe', 'Koisan', None),
 ('Zimbabwe', 'Zimbabwe', 'Kalanga', None),
 ('Zimbabwe', 'Zimbabwe', 'Chibarwe', None),
 ('Zimbabwe', 'Zimbabwe', 'Chewa', None),
 ('Zimbabwe', 'Zimbabwe', 'English', None),
 ('Zimbabwe', 'Zimbabwe', 'Ndebele', None),
 ('Zambia', 'Zambia', 'Lozi', 5.5),
 ('Zambia', 'Zambia', 'unspecified', 0.2),
 ('Zambia', 'Zambia', 'Other', 9.7),
 ('Zambia', 'Zambia', 'Bisa', 1.0),
 ('Zambia', 'Zambia', 'Lenje', 1.1),
 ('Zambia', 'Zambia', 'Namwanga', 1.2),
 ('Zambia', 'Zambia', 'Mambwe', 1.3),
 ('Zambia', 'Zambia', 'Luvale', 1.5),
 ('Zambia', 'Zambia', 'English', 1.7),
 ('Zambia', 'Zambia', 'Lamba', 1.8),
 ('Zambia', 'Zambia', 'Lala', 1.8),
 ('Zambia', 'Zambia', 'Kaonde', 1.8),
 ('Zambia', 'Zambia', 'Lunda', 1.9),
 ('Zambia', 'Zambia', 'Tumbuka', 2.5),
 ('Zambia', 'Zambia', 'Nsenga', 2.9),
 ('Zambia', 'Zambia', 'Chewa', 4.5),
 ('Zambia', 'Zambia', 'Tonga', 11.4),
 ('Zambia', 'Zambia', 'Nyanja', 14.7),
 ('Zambia', 'Zambia', 'Bembe', 33.4),
 ('Yemen', 'Al-Yaman', 'Arabic', None),
 ('Vietnam', 'Viet Nam', 'Other', None),
 ('Vietnam', 'Viet Nam', 'English', None),
 ('Vietnam', 'Viet Nam', 'Vietnamese', None),
 ('Venezuela', 'Venezuela', 'Spanish', None),
 ('Venezuela', 'Venezuela', 'indigenous', None),
 ('Vanuatu', 'Vanuatu', 'Other', 0.5),
 ('Vanuatu', 'Vanuatu', 'French', 0.6),
 ('Vanuatu', 'Vanuatu', 'English', 2.0),
 ('Vanuatu', 'Vanuatu', 'Bislama', 33.7),
 ('Vanuatu', 'Vanuatu', 'Tribal Languages', 63.2),
 ('Uzbekistan', 'Uzbekiston', 'Uzbek', 74.3),
 ('Uzbekistan', 'Uzbekiston', 'Russian', 14.2),
 ('Uzbekistan', 'Uzbekiston', 'Other', 7.1),
 ('Uzbekistan', 'Uzbekiston', 'Tajik', 4.4),
 ('Uruguay', 'Uruguay', 'Brazilero', None),
 ('Uruguay', 'Uruguay', 'Spanish', None),
 ('Uruguay', 'Uruguay', 'Portunol', None),
 ('United States', 'United States', 'Other', 8.0),
 ('United States', 'United States', 'Spanish', 12.9),
 ('United States', 'United States', 'English', 79.2),
 ('United Kingdom', 'United Kingdom', 'English', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Arabic', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Urdu', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Hindi', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'English', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Persian', None),
 ('Ukraine', 'Ukrajina', 'Russian', 29.6),
 ('Ukraine', 'Ukrajina', 'Other', 2.9),
 ('Ukraine', 'Ukrajina', 'Ukrainian', 67.5),
 ('Uganda', 'Uganda', 'Ganda', None),
 ('Uganda', 'Uganda', 'Arabic', None),
 ('Uganda', 'Uganda', 'Swahili', None),
 ('Uganda', 'Uganda', 'Nilo', None),
 ('Uganda', 'Uganda', 'Other', None),
 ('Uganda', 'Uganda', 'English', None),
 ('Tuvalu', 'Tuvalu', 'Tuvaluan', None),
 ('Tuvalu', 'Tuvalu', 'Kiribati', None),
 ('Tuvalu', 'Tuvalu', 'Samoan', None),
 ('Tuvalu', 'Tuvalu', 'English', None),
 ('Turks and Caicos Islands', 'The Turks and Caicos Islands', 'English', None),
 ('Turkmenistan', 'Turkmenostan', 'Russian', 12.0),
 ('Turkmenistan', 'Turkmenostan', 'Other', 7.0),
 ('Turkmenistan', 'Turkmenostan', 'Uzbek', 9.0),
 ('Turkmenistan', 'Turkmenostan', 'Turkmen', 72.0),
 ('Turkey', 'Turkiye', 'Kurdish', None),
 ('Turkey', 'Turkiye', 'Other', None),
 ('Turkey', 'Turkiye', 'Turkish', None),
 ('Tunisia', 'Tunis/Tunisie', 'French', None),
 ('Tunisia', 'Tunis/Tunisie', 'Arabic', None),
 ('Tunisia', 'Tunis/Tunisie', 'Berber', None),
 ('Trinidad and Tobago', 'Trinidad and Tobago', 'French', None),
 ('Trinidad and Tobago', 'Trinidad and Tobago', 'Engli

In [115]:
%%sql
select co.name as country, local_name,
       la.name as language, percent
from countries as co
    left join languages as la
        using(code)
order by country desc

 * postgresql://postgres:***@localhost:5432/postgres
921 rows affected.


[('Zimbabwe', 'Zimbabwe', 'Chibarwe', None),
 ('Zimbabwe', 'Zimbabwe', 'Shona', None),
 ('Zimbabwe', 'Zimbabwe', 'Ndebele', None),
 ('Zimbabwe', 'Zimbabwe', 'English', None),
 ('Zimbabwe', 'Zimbabwe', 'Chewa', None),
 ('Zimbabwe', 'Zimbabwe', 'Xhosa', None),
 ('Zimbabwe', 'Zimbabwe', 'Venda', None),
 ('Zimbabwe', 'Zimbabwe', 'Tswana', None),
 ('Zimbabwe', 'Zimbabwe', 'Tonga', None),
 ('Zimbabwe', 'Zimbabwe', 'Sotho', None),
 ('Zimbabwe', 'Zimbabwe', 'sign', None),
 ('Zimbabwe', 'Zimbabwe', 'Shangani', None),
 ('Zimbabwe', 'Zimbabwe', 'Ndau', None),
 ('Zimbabwe', 'Zimbabwe', 'Nambya', None),
 ('Zimbabwe', 'Zimbabwe', 'Koisan', None),
 ('Zimbabwe', 'Zimbabwe', 'Kalanga', None),
 ('Zambia', 'Zambia', 'Lala', 1.8),
 ('Zambia', 'Zambia', 'Mambwe', 1.3),
 ('Zambia', 'Zambia', 'Luvale', 1.5),
 ('Zambia', 'Zambia', 'English', 1.7),
 ('Zambia', 'Zambia', 'Lamba', 1.8),
 ('Zambia', 'Zambia', 'Kaonde', 1.8),
 ('Zambia', 'Zambia', 'Lunda', 1.9),
 ('Zambia', 'Zambia', 'Tumbuka', 2.5),
 ('Zambia', 'Zambia', 'Nsenga', 2.9),
 ('Zambia', 'Zambia', 'Chewa', 4.5),
 ('Zambia', 'Zambia', 'Lozi', 5.5),
 ('Zambia', 'Zambia', 'Tonga', 11.4),
 ('Zambia', 'Zambia', 'Nyanja', 14.7),
 ('Zambia', 'Zambia', 'Bembe', 33.4),
 ('Zambia', 'Zambia', 'unspecified', 0.2),
 ('Zambia', 'Zambia', 'Other', 9.7),
 ('Zambia', 'Zambia', 'Bisa', 1.0),
 ('Zambia', 'Zambia', 'Lenje', 1.1),
 ('Zambia', 'Zambia', 'Namwanga', 1.2),
 ('Yemen', 'Al-Yaman', 'Arabic', None),
 ('Virgin Islands, U.S.', 'Virgin Islands of the United States', None, None),
 ('Virgin Islands, British', 'British Virgin Islands', None, None),
 ('Vietnam', 'Viet Nam', 'Other', None),
 ('Vietnam', 'Viet Nam', 'Vietnamese', None),
 ('Vietnam', 'Viet Nam', 'English', None),
 ('Venezuela', 'Venezuela', 'indigenous', None),
 ('Venezuela', 'Venezuela', 'Spanish', None),
 ('Vanuatu', 'Vanuatu', 'Other', 0.5),
 ('Vanuatu', 'Vanuatu', 'Tribal Languages', 63.2),
 ('Vanuatu', 'Vanuatu', 'English', 2.0),
 ('Vanuatu', 'Vanuatu', 'French', 0.6),
 ('Vanuatu', 'Vanuatu', 'Bislama', 33.7),
 ('Uzbekistan', 'Uzbekiston', 'Uzbek', 74.3),
 ('Uzbekistan', 'Uzbekiston', 'Other', 7.1),
 ('Uzbekistan', 'Uzbekiston', 'Tajik', 4.4),
 ('Uzbekistan', 'Uzbekiston', 'Russian', 14.2),
 ('Uruguay', 'Uruguay', 'Brazilero', None),
 ('Uruguay', 'Uruguay', 'Portunol', None),
 ('Uruguay', 'Uruguay', 'Spanish', None),
 ('United States', 'United States', 'Spanish', 12.9),
 ('United States', 'United States', 'Other', 8.0),
 ('United States', 'United States', 'English', 79.2),
 ('United Kingdom', 'United Kingdom', 'English', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Arabic', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Hindi', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'English', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Persian', None),
 ('United Arab Emirates', 'Al-Imarat al-´Arabiya al-Muttahida', 'Urdu', None),
 ('Ukraine', 'Ukrajina', 'Ukrainian', 67.5),
 ('Ukraine', 'Ukrajina', 'Russian', 29.6),
 ('Ukraine', 'Ukrajina', 'Other', 2.9),
 ('Uganda', 'Uganda', 'Nilo', None),
 ('Uganda', 'Uganda', 'Ganda', None),
 ('Uganda', 'Uganda', 'English', None),
 ('Uganda', 'Uganda', 'Arabic', None),
 ('Uganda', 'Uganda', 'Swahili', None),
 ('Uganda', 'Uganda', 'Other', None),
 ('Tuvalu', 'Tuvalu', 'Kiribati', None),
 ('Tuvalu', 'Tuvalu', 'Tuvaluan', None),
 ('Tuvalu', 'Tuvalu', 'English', None),
 ('Tuvalu', 'Tuvalu', 'Samoan', None),
 ('Turks and Caicos Islands', 'The Turks and Caicos Islands', 'English', None),
 ('Turkmenistan', 'Turkmenostan', 'Uzbek', 9.0),
 ('Turkmenistan', 'Turkmenostan', 'Turkmen', 72.0),
 ('Turkmenistan', 'Turkmenostan', 'Russian', 12.0),
 ('Turkmenistan', 'Turkmenostan', 'Other', 7.0),
 ('Turkey', 'Turkiye', 'Turkish', None),
 ('Turkey', 'Turkiye', 'Other', None),
 ('Turkey', 'Turkiye', 'Kurdish', None),
 ('Tunisia', 'Tunis/Tunisie', 'Berber', None),
 ('Tunisia', 'Tunis/Tunisie', 'Arabic', None),
 ('Tunisia', 'Tunis/

**Exercise**<br>
* Determine the average gross domestic product (GDP) per capita by region in 2010.

In [119]:
%%sql
select region, AVG(gdp_percapita) as avg_gdp_percapita
from countries
    left join economies
        using(code)
where year = 2010
group by region
order by avg_gdp_percapita desc

 * postgresql://postgres:***@localhost:5432/postgres
23 rows affected.


[('Western Europe', 58130.9614955357),
 ('Nordic Countries', 57073.99765625),
 ('North America', 47911.509765625),
 ('Australia and New Zealand', 44792.384765625),
 ('British Islands', 43588.330078125),
 ('Eastern Asia', 26205.8513997396),
 ('Southern Europe', 22926.4109108665),
 ('Middle East', 18204.6415153952),
 ('Baltic Countries', 12631.0299479167),
 ('Caribbean', 11413.339454064),
 ('Southeast Asia', 10547.1541320801),
 ('Eastern Europe', 10095.4565700955),
 ('South America', 7420.59923299154),
 ('Southern Africa', 5051.59797363281),
 ('Central America', 4969.97006225586),
 ('Central Africa', 4797.23992919922),
 ('Northern Africa', 4713.86169433594),
 ('Micronesia', 4497.65197753906),
 ('Polynesia', 3451.36336263021),
 ('Southern and Central Asia', 2797.15499441964),
 ('Melanesia', 2532.61001586914),
 ('Eastern Africa', 1757.34816169739),
 ('Western Africa', 998.954866409302)]

## RIGHT JOIN
<code>RIGHT JOIN</code> is the mirror of the left join.
<img src="images/right_join.png" alt='Drawing' style="width: 200px;"/>

**Exercise**<br>
Convert this code to use RIGHT JOINs instead of LEFT JOINs:
<code>
SELECT cities.name AS city, urbanarea_pop, countries.name AS country,
       indep_year, languages.name AS language, percent
FROM cities
LEFT JOIN countries
ON cities.country_code = countries.code
LEFT JOIN languages
ON countries.code = languages.code
ORDER BY city, language;
</code>

In [120]:
%%sql
SELECT cities.name AS city, urbanarea_pop, countries.name AS country,
       indep_year, languages.name AS language, percent
FROM cities
LEFT JOIN countries
ON cities.country_code = countries.code
LEFT JOIN languages
ON countries.code = languages.code
ORDER BY city, language
limit 5

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('Abidjan', 4765000.0, "Cote d'Ivoire", 1960, 'French', None),
 ('Abidjan', 4765000.0, "Cote d'Ivoire", 1960, 'Other', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'Arabic', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'English', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'Hindi', None)]

In [121]:
%%sql
select ci.name as city, urbanarea_pop,
       co.name as country, indep_year,
       la.name as language, percent
from languages as la
    right join countries as co
        using(code)
    right join cities as ci
        on ci.country_code=la.code
order by city, language
limit 5

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('Abidjan', 4765000.0, "Cote d'Ivoire", 1960, 'French', None),
 ('Abidjan', 4765000.0, "Cote d'Ivoire", 1960, 'Other', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'Arabic', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'English', None),
 ('Abu Dhabi', 1145000.0, 'United Arab Emirates', 1971, 'Hindi', None)]

## FULL JOIN
<img src="images/full_join.png" alt='Drawing' style="width: 200px;"/>

**Exercise**<br>
* Return country name, code, region and basic_unit for North America or where it is missing. Order by region.

In [124]:
%%sql
select name as country, code, region, basic_unit
from countries
    full join currencies
        using(code)
where region = 'North America' or region is null
order by region

 * postgresql://postgres:***@localhost:5432/postgres
17 rows affected.


[('Canada', 'CAN', 'North America', 'Canadian dollar'),
 ('United States', 'USA', 'North America', 'United States dollar'),
 ('Bermuda', 'BMU', 'North America', 'Bermudian dollar'),
 ('Greenland', 'GRL', 'North America', None),
 (None, 'TMP', None, 'United States dollar'),
 (None, 'FLK', None, 'Falkland Islands pound'),
 (None, 'AIA', None, 'East Caribbean dollar'),
 (None, 'NIU', None, 'New Zealand dollar'),
 (None, 'ROM', None, 'Romanian leu'),
 (None, 'SHN', None, 'Saint Helena pound'),
 (None, 'SGS', None, 'British pound'),
 (None, 'TWN', None, 'New Taiwan dollar'),
 (None, 'WLF', None, 'CFP franc'),
 (None, 'MSR', None, 'East Caribbean dollar'),
 (None, 'IOT', None, 'United States dollar'),
 (None, 'CCK', None, 'Australian dollar'),
 (None, 'COK', None, 'New Zealand dollar')]

**Exercise**<br>
1. Full join with *languages* on the left and *countries* on the right.
2. Select country, code, language.
3. Country starts with 'V' or is NULL
4. Order by country ascendingly

In this Ex languages has 955 records vs 206 records for countries

In [128]:
%%sql
select co.name as country, la.code, la.name as language
from languages as la
    full join countries as co
        using(code)
where co.name like 'V%' or co.name is null
order by country

 * postgresql://postgres:***@localhost:5432/postgres
53 rows affected.


[('Vanuatu', 'VUT', 'Tribal Languages'),
 ('Vanuatu', 'VUT', 'English'),
 ('Vanuatu', 'VUT', 'French'),
 ('Vanuatu', 'VUT', 'Other'),
 ('Vanuatu', 'VUT', 'Bislama'),
 ('Venezuela', 'VEN', 'Spanish'),
 ('Venezuela', 'VEN', 'indigenous'),
 ('Vietnam', 'VNM', 'Vietnamese'),
 ('Vietnam', 'VNM', 'English'),
 ('Vietnam', 'VNM', 'Other'),
 ('Virgin Islands, British', None, None),
 ('Virgin Islands, U.S.', None, None),
 (None, 'NIU', 'English'),
 (None, 'NIU', 'Niuean'),
 (None, 'NIU', 'Other'),
 (None, 'NFK', 'English'),
 (None, 'NFK', 'Other'),
 (None, 'ROM', 'Romanian'),
 (None, 'ROM', 'Hungarian'),
 (None, 'ROM', 'Romani'),
 (None, 'ROM', 'Other'),
 (None, 'ROM', 'unspecified'),
 (None, 'SPM', 'French'),
 (None, 'TWN', 'Mandarin'),
 (None, 'TWN', 'Taiwanese'),
 (None, 'TWN', 'Hakka'),
 (None, 'TKL', 'Tokelauan'),
 (None, 'TKL', 'English'),
 (None, 'TKL', 'Samoan'),
 (None, 'TKL', 'Tuvaluan'),
 (None, 'TKL', 'Kiribati'),
 (None, 'TKL', 'Other'),
 (None, 'AIA', 'English'),
 (None, 'TKL', 'unspecified'),
 (None, 'WLF', 'Wallisian'),
 (None, 'WLF', 'Futunian'),
 (None, 'WLF', 'French'),
 (None, 'WLF', 'Other'),
 (None, 'ESH', 'Standard'),
 (None, 'ESH', 'Hassaniya'),
 (None, 'ESH', 'Moroccan'),
 (None, 'TKL', 'none'),
 (None, 'CXR', 'English'),
 (None, 'CXR', 'Chinese'),
 (None, 'CXR', 'Malay'),
 (None, 'CCK', 'Malay'),
 (None, 'CCK', 'English'),
 (None, 'COK', 'English'),
 (None, 'COK', 'Rarotongan'),
 (None, 'COK', 'Other'),
 (None, 'MSR', 'English'),
 (None, 'NIU', 'Niuean'),
 (None, 'NIU', 'Niuean')]

**Ecxercise**<br>
1. Complete a full join with countries on the left and languages on the right. Select country, region, language, basic_unit and frac_unit.
2. Next, full join this result with currencies on the right. Choose the Melanesia and Micronesia regions.

In [129]:
%%sql
select co.name as country, region, 
       la.name as language,
       basic_unit, frac_unit
from countries as co
    full join languages as la
        using(code)
    full join currencies
        using(code)
where region like 'M%nesia'

 * postgresql://postgres:***@localhost:5432/postgres
50 rows affected.


[('Kiribati', 'Micronesia', 'English', 'Australian dollar', 'Cent'),
 ('Kiribati', 'Micronesia', 'Kiribati', 'Australian dollar', 'Cent'),
 ('Marshall Islands', 'Micronesia', 'Other', 'United States dollar', 'Cent'),
 ('Marshall Islands', 'Micronesia', 'Marshallese', 'United States dollar', 'Cent'),
 ('Nauru', 'Micronesia', 'Other', 'Australian dollar', 'Cent'),
 ('Nauru', 'Micronesia', 'English', 'Australian dollar', 'Cent'),
 ('Nauru', 'Micronesia', 'Nauruan', 'Australian dollar', 'Cent'),
 ('New Caledonia', 'Melanesia', 'Other', 'CFP franc', 'Centime'),
 ('New Caledonia', 'Melanesia', 'French', 'CFP franc', 'Centime'),
 ('Palau', 'Micronesia', 'Other', 'United States dollar', 'Cent'),
 ('Palau', 'Micronesia', 'Chinese', 'United States dollar', 'Cent'),
 ('Palau', 'Micronesia', 'Filipino', 'United States dollar', 'Cent'),
 ('Palau', 'Micronesia', 'English', 'United States dollar', 'Cent'),
 ('Palau', 'Micronesia', 'Other', 'United States dollar', 'Cent'),
 ('Palau', 'Micronesia', 'Palauan', 'United States dollar', 'Cent'),
 ('Papua New Guinea', 'Melanesia', 'Other', 'Papua New Guinean kina', 'Toea'),
 ('Papua New Guinea', 'Melanesia', 'Hiri', 'Papua New Guinean kina', 'Toea'),
 ('Papua New Guinea', 'Melanesia', 'English', 'Papua New Guinean kina', 'Toea'),
 ('Papua New Guinea', 'Melanesia', 'Tok Pisin', 'Papua New Guinean kina', 'Toea'),
 ('Solomon Islands', 'Melanesia', 'indigenous', 'Solomon Islands dollar', 'Cent'),
 ('Solomon Islands', 'Melanesia', 'English', 'Solomon Islands dollar', 'Cent'),
 ('Solomon Islands', 'Melanesia', 'Melanesian pidgin', 'Solomon Islands dollar', 'Cent'),
 ('Vanuatu', 'Melanesia', 'Other', 'Vanuatu vatu', None),
 ('Vanuatu', 'Melanesia', 'French', 'Vanuatu vatu', None),
 ('Vanuatu', 'Melanesia', 'English', 'Vanuatu vatu', None),
 ('Vanuatu', 'Melanesia', 'Bislama', 'Vanuatu vatu', None),
 ('Vanuatu', 'Melanesia', 'Tribal Languages', 'Vanuatu vatu', None),
 ('Micronesia, Federated States of', 'Micronesia', 'Kapingamarangi', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Nukuoro', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Woleaian', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Ulithian', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Yapese', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Pohnpeian', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Kosrean', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'Chuukese', None, None),
 ('Micronesia, Federated States of', 'Micronesia', 'English', None, None),
 ('Fiji Islands', 'Melanesia', None, None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Other', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Other Asian', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Chinese', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Other Pacific Island', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'English', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Chamorro', None, None),
 ('Northern Mariana Islands', 'Micronesia', 'Philippine', None, None),
 ('Guam', 'Micronesia', 'Other', None, None),
 ('Guam', 'Micronesia', 'Asian', None, None),
 ('Guam', 'Micronesia', 'Other Pacific Islander', None, None),
 ('Guam', 'Micronesia', 'Chamorro', None, None),
 ('Guam', 'Micronesia', 'Filipino', None, None),
 ('Guam', 'Micronesia', 'English', None, None)]

## CROSS JOIN
Creates all possible combinations of joins.
<img src="images/cross_join.png" alt='Drawing' style="width: 100px;"/>

**Exercise**
1. Explore languages potentially and most frequently spoken in the cities of Hyderabad, India and Hyderabad, Pakistan.

In [131]:
%%sql
select ci.name as city, la.name as language
from cities as ci
    cross join languages as la
where ci.name like 'Hyder%'

 * postgresql://postgres:***@localhost:5432/postgres
1910 rows affected.


[('Hyderabad (India)', 'Dari'),
 ('Hyderabad', 'Dari'),
 ('Hyderabad (India)', 'Pashto'),
 ('Hyderabad', 'Pashto'),
 ('Hyderabad (India)', 'Turkic'),
 ('Hyderabad', 'Turkic'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'Albanian'),
 ('Hyderabad', 'Albanian'),
 ('Hyderabad (India)', 'Greek'),
 ('Hyderabad', 'Greek'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'unspecified'),
 ('Hyderabad', 'unspecified'),
 ('Hyderabad (India)', 'Arabic'),
 ('Hyderabad', 'Arabic'),
 ('Hyderabad (India)', 'French'),
 ('Hyderabad', 'French'),
 ('Hyderabad (India)', 'Berber or Tamazight'),
 ('Hyderabad', 'Berber or Tamazight'),
 ('Hyderabad (India)', 'Shawiya'),
 ('Hyderabad', 'Shawiya'),
 ('Hyderabad (India)', 'Mzab'),
 ('Hyderabad', 'Mzab'),
 ('Hyderabad (India)', 'Tuareg'),
 ('Hyderabad', 'Tuareg'),
 ('Hyderabad (India)', 'Samoan'),
 ('Hyderabad', 'Samoan'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'Tongan'),
 ('Hyderabad', 'Tongan'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'Catalan'),
 ('Hyderabad', 'Catalan'),
 ('Hyderabad (India)', 'French'),
 ('Hyderabad', 'French'),
 ('Hyderabad (India)', 'Castilian'),
 ('Hyderabad', 'Castilian'),
 ('Hyderabad (India)', 'Portuguese'),
 ('Hyderabad', 'Portuguese'),
 ('Hyderabad (India)', 'Portuguese'),
 ('Hyderabad', 'Portuguese'),
 ('Hyderabad (India)', 'Umbundu'),
 ('Hyderabad', 'Umbundu'),
 ('Hyderabad (India)', 'Kikongo'),
 ('Hyderabad', 'Kikongo'),
 ('Hyderabad (India)', 'Kimbundu'),
 ('Hyderabad', 'Kimbundu'),
 ('Hyderabad (India)', 'Chokwe'),
 ('Hyderabad', 'Chokwe'),
 ('Hyderabad (India)', 'Nhaneca'),
 ('Hyderabad', 'Nhaneca'),
 ('Hyderabad (India)', 'Nganguela'),
 ('Hyderabad', 'Nganguela'),
 ('Hyderabad (India)', 'Fiote'),
 ('Hyderabad', 'Fiote'),
 ('Hyderabad (India)', 'Kwanhama'),
 ('Hyderabad', 'Kwanhama'),
 ('Hyderabad (India)', 'Muhumbi'),
 ('Hyderabad', 'Muhumbi'),
 ('Hyderabad (India)', 'Luvale'),
 ('Hyderabad', 'Luvale'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'Antiguan creole'),
 ('Hyderabad', 'Antiguan creole'),
 ('Hyderabad (India)', 'Spanish'),
 ('Hyderabad', 'Spanish'),
 ('Hyderabad (India)', 'Italian'),
 ('Hyderabad', 'Italian'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'German'),
 ('Hyderabad', 'German'),
 ('Hyderabad (India)', 'French'),
 ('Hyderabad', 'French'),
 ('Hyderabad (India)', 'indigenous'),
 ('Hyderabad', 'indigenous'),
 ('Hyderabad (India)', 'Armenian'),
 ('Hyderabad', 'Armenian'),
 ('Hyderabad (India)', 'Kurdish'),
 ('Hyderabad', 'Kurdish'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'Papiamento'),
 ('Hyderabad', 'Papiamento'),
 ('Hyderabad (India)', 'Spanish'),
 ('Hyderabad', 'Spanish'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'Dutch'),
 ('Hyderabad', 'Dutch'),
 ('Hyderabad (India)', 'Chinese'),
 ('Hyderabad', 'Chinese'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'unspecified'),
 ('Hyderabad', 'unspecified'),
 ('Hyderabad (India)', 'English'),
 ('Hyderabad', 'English'),
 ('Hyderabad (India)', 'Mandarin'),
 ('Hyderabad', 'Mandarin'),
 ('Hyderabad (India)', 'Italian'),
 ('Hyderabad', 'Italian'),
 ('Hyderabad (India)', 'Arabic'),
 ('Hyderabad', 'Arabic'),
 ('Hyderabad (India)', 'Greek'),
 ('Hyderabad', 'Greek'),
 ('Hyderabad (India)', 'Cantonese'),
 ('Hyderabad', 'Cantonese'),
 ('Hyderabad (India)', 'Vietnamese'),
 ('Hyderabad', 'Vietnamese'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Other'),
 ('Hyderabad (India)', 'unspecified'),
 ('Hyderabad', 'unspecified'),
 ('Hyderabad (India)', 'German'),
 ('Hyderabad', 'German'),
 ('H

In [132]:
%%sql
--for comparison to cross join we look at inner join of the same query
select ci.name as city, la.name as language
from cities as ci
    inner join languages as la
        on ci.country_code = la.code
where ci.name like 'Hyder%'

 * postgresql://postgres:***@localhost:5432/postgres
25 rows affected.


[('Hyderabad (India)', 'Hindi'),
 ('Hyderabad (India)', 'Bengali'),
 ('Hyderabad (India)', 'Telugu'),
 ('Hyderabad (India)', 'Marathi'),
 ('Hyderabad (India)', 'Tamil'),
 ('Hyderabad (India)', 'Urdu'),
 ('Hyderabad (India)', 'Gujarati'),
 ('Hyderabad (India)', 'Kannada'),
 ('Hyderabad (India)', 'Malayalam'),
 ('Hyderabad (India)', 'Oriya'),
 ('Hyderabad (India)', 'Punjabi'),
 ('Hyderabad (India)', 'Assamese'),
 ('Hyderabad (India)', 'Maithili'),
 ('Hyderabad (India)', 'Other'),
 ('Hyderabad', 'Punjabi'),
 ('Hyderabad', 'Sindhi'),
 ('Hyderabad', 'Saraiki'),
 ('Hyderabad', 'Pashto'),
 ('Hyderabad', 'Urdu'),
 ('Hyderabad', 'Balochi'),
 ('Hyderabad', 'Hindko'),
 ('Hyderabad', 'Brahui'),
 ('Hyderabad', 'English'),
 ('Hyderabad', 'Burushaski'),
 ('Hyderabad', 'Other')]

**Exercise**<br>
In terms of life expectancy for 2010, determine the names of the lowest five countries and their regions.

In [135]:
%%sql
select co.name as country, region,
       life_expectancy
from populations as po
    left join countries as co
        on po.country_code = co.code
where year = 2010
order by life_expectancy
limit 5

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('Lesotho', 'Southern Africa', 47.4834),
 ('Central African Republic', 'Central Africa', 47.6253),
 ('Sierra Leone', 'Western Africa', 48.229),
 ('Swaziland', 'Southern Africa', 48.3458),
 ('Zimbabwe', 'Eastern Africa', 49.5747)]

## Semi-join
<img src="images/semi_join.png" style="width: 200px;"/>

**Exercise**<br>
*  Identify languages spoken in the Middle East.

In [164]:
%%sql
select distinct(name)
from languages
where code in 
    (select code
     from countries
     where region like 'Middle East')
order by name

 * postgresql://postgres:***@localhost:5432/postgres
27 rows affected.


[('Arabic',),
 ('Aramaic',),
 ('Armenian',),
 ('Azerbaijani',),
 ('Azeri',),
 ('Baluchi',),
 ('Bulgarian',),
 ('Circassian',),
 ('English',),
 ('Farsi',),
 ('Filipino',),
 ('French',),
 ('Georgian',),
 ('Greek',),
 ('Hebrew',),
 ('Hindi',),
 ('Indian',),
 ('Kurdish',),
 ('Other',),
 ('Persian',),
 ('Romanian',),
 ('Russian',),
 ('Syriac',),
 ('Turkish',),
 ('Turkmen',),
 ('unspecified',),
 ('Urdu',)]

In [174]:
%%sql
--doing it via inner join
select distinct(languages.name)
from languages
    inner join countries
        using(code)
where countries.region like 'Middle East'
order by languages.name

 * postgresql://postgres:***@localhost:5432/postgres
27 rows affected.


[('Arabic',),
 ('Aramaic',),
 ('Armenian',),
 ('Azerbaijani',),
 ('Azeri',),
 ('Baluchi',),
 ('Bulgarian',),
 ('Circassian',),
 ('English',),
 ('Farsi',),
 ('Filipino',),
 ('French',),
 ('Georgian',),
 ('Greek',),
 ('Hebrew',),
 ('Hindi',),
 ('Indian',),
 ('Kurdish',),
 ('Other',),
 ('Persian',),
 ('Romanian',),
 ('Russian',),
 ('Syriac',),
 ('Turkish',),
 ('Turkmen',),
 ('unspecified',),
 ('Urdu',)]

## ANTI JOIN
<img src="images/anti_join.png" alt='Drawing' style="width: 200px;"/>

**Exercise**<br>
* identify the currencies used in Oceanian countries

In [182]:
%%sql
select co.code, name, basic_unit as currency
from countries as co
    inner join currencies
        using(code)
where continent='Oceania'

 * postgresql://postgres:***@localhost:5432/postgres
15 rows affected.


[('AUS', 'Australia', 'Australian dollar'),
 ('PYF', 'French Polynesia', 'CFP franc'),
 ('KIR', 'Kiribati', 'Australian dollar'),
 ('MHL', 'Marshall Islands', 'United States dollar'),
 ('NRU', 'Nauru', 'Australian dollar'),
 ('NCL', 'New Caledonia', 'CFP franc'),
 ('NZL', 'New Zealand', 'New Zealand dollar'),
 ('PLW', 'Palau', 'United States dollar'),
 ('PNG', 'Papua New Guinea', 'Papua New Guinean kina'),
 ('WSM', 'Samoa', 'Samoan tala'),
 ('SLB', 'Solomon Islands', 'Solomon Islands dollar'),
 ('TON', 'Tonga', 'Tongan paʻanga'),
 ('TUV', 'Tuvalu', 'Australian dollar'),
 ('TUV', 'Tuvalu', 'Tuvaluan dollar'),
 ('VUT', 'Vanuatu', 'Vanuatu vatu')]

In [183]:
%%sql
--number of countries in Oceania
select count(*)
from countries
where continent = 'Oceania'

 * postgresql://postgres:***@localhost:5432/postgres
1 rows affected.


[(19,)]

Not all countries in Oceania were listed in the resulting inner join with currencies. Use an anti-join to determine which countries were not included!

In [189]:
%%sql
select name
from countries
where continent='Oceania' and code not in
    (select code
     from currencies)

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('American Samoa',),
 ('Fiji Islands',),
 ('Guam',),
 ('Micronesia, Federated States of',),
 ('Northern Mariana Islands',)]

# Sets
Venn Diagrams:
<img src="images/venn_diagrams.png" style="width: 300px;"/>
where shaded area shows what is **included**.<br>
* UNION - Returns records in both tables.
* UNION ALL - Returns all records from either table and double returns records that are in both tables.
* INTERSECT - Returns records that are only in both tables.
* EXCEPT - Returns records only in one and not the other excluding matching records.

## UNION
<img src="images/union_dia.png" style="width: 100px;"/>
The table form looks like this:
<img src="images/union_table.png" style="width: 150px;"/>
The ID values of 1 and 4 from <code>right_table</code> are not included to the result since they were already found in the first table.

**Exercise**<br>
1. Combine <code>economies2010</code> and <code>economies2015</code> tables containing all of the fields in <code>economies2010</code>.
2. Sort this resulting single table by country code and then by year, both in ascending order.

In [141]:
%%sql
select *
from economies2010
UNION
select *
from economies2015
order by code,year
limit 6

 * postgresql://postgres:***@localhost:5432/postgres
6 rows affected.


[('AFG', 2010, 'Low income', 37.133),
 ('AFG', 2015, 'Low income', 21.466),
 ('AGO', 2010, 'Upper middle income', 23.534),
 ('AGO', 2015, 'Upper middle income', -0.425),
 ('ALB', 2010, 'Upper middle income', 20.011),
 ('ALB', 2015, 'Upper middle income', 13.84)]

In [144]:
%%sql
--as a refernece let's look at economies 6 first rows
select code,year,income_group, gross_savings
from economies
order by code, year
limit 6

 * postgresql://postgres:***@localhost:5432/postgres
6 rows affected.


[('AFG', 2010, 'Low income', 37.133),
 ('AFG', 2015, 'Low income', 21.466),
 ('AGO', 2010, 'Upper middle income', 23.534),
 ('AGO', 2015, 'Upper middle income', -0.425),
 ('ALB', 2010, 'Upper middle income', 20.011),
 ('ALB', 2015, 'Upper middle income', 13.84)]

**Exercise**<br>
1. Determine all (non-duplicated) country codes in either the cities or the currencies table. The result should be a table with only one field called country_code.
2. Sort by country_code in alphabetical order.

In [148]:
%%sql
select country_code
from cities
UNION
select code
from currencies
order by country_code
limit 5

 * postgresql://postgres:***@localhost:5432/postgres
5 rows affected.


[('ABW',), ('AFG',), ('AGO',), ('AIA',), ('ALB',)]

## UNION ALL
<img src="images/unionall_dia.png" style="width: 100px;"/>
The table form looks like this:
<img src="images/unionall_table.png" style="width: 150px;"/>

**Exercise**<br>
1. Determine all (include duplicates) country codes and years that exist in either the economies or the populations tables. Order by code then year.

In [150]:
%%sql
select code, year
from economies
UNION ALL 
select country_code, year
from populations
order by code, year

 * postgresql://postgres:***@localhost:5432/postgres
814 rows affected.


[('ABW', 2010),
 ('ABW', 2015),
 ('AFG', 2010),
 ('AFG', 2010),
 ('AFG', 2015),
 ('AFG', 2015),
 ('AGO', 2010),
 ('AGO', 2010),
 ('AGO', 2015),
 ('AGO', 2015),
 ('ALB', 2010),
 ('ALB', 2010),
 ('ALB', 2015),
 ('ALB', 2015),
 ('AND', 2010),
 ('AND', 2015),
 ('ARE', 2010),
 ('ARE', 2010),
 ('ARE', 2015),
 ('ARE', 2015),
 ('ARG', 2010),
 ('ARG', 2010),
 ('ARG', 2015),
 ('ARG', 2015),
 ('ARM', 2010),
 ('ARM', 2010),
 ('ARM', 2015),
 ('ARM', 2015),
 ('ASM', 2010),
 ('ASM', 2015),
 ('ATG', 2010),
 ('ATG', 2010),
 ('ATG', 2015),
 ('ATG', 2015),
 ('AUS', 2010),
 ('AUS', 2010),
 ('AUS', 2015),
 ('AUS', 2015),
 ('AUT', 2010),
 ('AUT', 2010),
 ('AUT', 2015),
 ('AUT', 2015),
 ('AZE', 2010),
 ('AZE', 2010),
 ('AZE', 2015),
 ('AZE', 2015),
 ('BDI', 2010),
 ('BDI', 2010),
 ('BDI', 2015),
 ('BDI', 2015),
 ('BEL', 2010),
 ('BEL', 2010),
 ('BEL', 2015),
 ('BEL', 2015),
 ('BEN', 2010),
 ('BEN', 2010),
 ('BEN', 2015),
 ('BEN', 2015),
 ('BFA', 2010),
 ('BFA', 2010),
 ('BFA', 2015),
 ('BFA', 2015),
 ('BGD', 2010),
 ('BGD', 2010),
 ('BGD', 2015),
 ('BGD', 2015),
 ('BGR', 2010),
 ('BGR', 2010),
 ('BGR', 2015),
 ('BGR', 2015),
 ('BHR', 2010),
 ('BHR', 2010),
 ('BHR', 2015),
 ('BHR', 2015),
 ('BHS', 2010),
 ('BHS', 2010),
 ('BHS', 2015),
 ('BHS', 2015),
 ('BIH', 2010),
 ('BIH', 2010),
 ('BIH', 2015),
 ('BIH', 2015),
 ('BLR', 2010),
 ('BLR', 2010),
 ('BLR', 2015),
 ('BLR', 2015),
 ('BLZ', 2010),
 ('BLZ', 2010),
 ('BLZ', 2015),
 ('BLZ', 2015),
 ('BMU', 2010),
 ('BMU', 2015),
 ('BOL', 2010),
 ('BOL', 2010),
 ('BOL', 2015),
 ('BOL', 2015),
 ('BRA', 2010),
 ('BRA', 2010),
 ('BRA', 2015),
 ('BRA', 2015),
 ('BRB', 2010),
 ('BRB', 2010),
 ('BRB', 2015),
 ('BRB', 2015),
 ('BRN', 2010),
 ('BRN', 2010),
 ('BRN', 2015),
 ('BRN', 2015),
 ('BTN', 2010),
 ('BTN', 2010),
 ('BTN', 2015),
 ('BTN', 2015),
 ('BWA', 2010),
 ('BWA', 2010),
 ('BWA', 2015),
 ('BWA', 2015),
 ('CAF', 2010),
 ('CAF', 2010),
 ('CAF', 2015),
 ('CAF', 2015),
 ('CAN', 2010),
 ('CAN', 2010),
 ('CAN', 2015),
 ('CAN', 2015),
 ('CHE', 2010),
 ('CHE', 2010),
 ('CHE', 2015),
 ('CHE', 2015),
 ('CHI', 2010),
 ('CHI', 2015),
 ('CHL', 2010),
 ('CHL', 2010),
 ('CHL', 2015),
 ('CHL', 2015),
 ('CHN', 2010),
 ('CHN', 2010),
 ('CHN', 2015),
 ('CHN', 2015),
 ('CIV', 2010),
 ('CIV', 2010),
 ('CIV', 2015),
 ('CIV', 2015),
 ('CMR', 2010),
 ('CMR', 2010),
 ('CMR', 2015),
 ('CMR', 2015),
 ('COD', 2010),
 ('COD', 2010),
 ('COD', 2015),
 ('COD', 2015),
 ('COG', 2010),
 ('COG', 2010),
 ('COG', 2015),
 ('COG', 2015),
 ('COL', 2010),
 ('COL', 2010),
 ('COL', 2015),
 ('COL', 2015),
 ('COM', 2010),
 ('COM', 2010),
 ('COM', 2015),
 ('COM', 2015),
 ('CPV', 2010),
 ('CPV', 2010),
 ('CPV', 2015),
 ('CPV', 2015),
 ('CRI', 2010),
 ('CRI', 2010),
 ('CRI', 2015),
 ('CRI', 2015),
 ('CUB', 2010),
 ('CUB', 2015),
 ('CUW', 2010),
 ('CUW', 2015),
 ('CYM', 2010),
 ('CYM', 2015),
 ('CYP', 2010),
 ('CYP', 2010),
 ('CYP', 2015),
 ('CYP', 2015),
 ('CZE', 2010),
 ('CZE', 2010),
 ('CZE', 2015),
 ('CZE', 2015),
 ('DEU', 2010),
 ('DEU', 2010),
 ('DEU', 2015),
 ('DEU', 2015),
 ('DJI', 2010),
 ('DJI', 2010),
 ('DJI', 2015),
 ('DJI', 2015),
 ('DMA', 2010),
 ('DMA', 2010),
 ('DMA', 2015),
 ('DMA', 2015),
 ('DNK', 2010),
 ('DNK', 2010),
 ('DNK', 2015),
 ('DNK', 2015),
 ('DOM', 2010),
 ('DOM', 2010),
 ('DOM', 2015),
 ('DOM', 2015),
 ('DZA', 2010),
 ('DZA', 2010),
 ('DZA', 2015),
 ('DZA', 2015),
 ('ECU', 2010),
 ('ECU', 2010),
 ('ECU', 2015),
 ('ECU', 2015),
 ('EGY', 2010),
 ('EGY', 2010),
 ('EGY', 2015),
 ('EGY', 2015),
 ('ERI', 2010),
 ('ERI', 2010),
 ('ERI', 2015),
 ('ERI', 2015),
 ('ESP', 2010),
 ('ESP', 2010),
 ('ESP', 2015),
 ('ESP', 2015),
 ('EST', 2010),
 ('EST', 2010),
 ('EST', 2015),
 ('EST', 2015),
 ('ETH', 2010),
 ('ETH', 2010),
 ('ETH', 2015),
 ('ETH', 2015),
 ('FIN', 2010),
 ('FIN', 2010),
 ('FIN', 2015),
 ('FIN', 2015),
 ('FJI', 2010),
 ('FJI', 2010),
 ('FJI', 2015),
 ('FJI', 2015),
 ('FRA', 2010),
 ('FRA', 2010),
 ('FRA', 2015),
 ('FRA', 2015),
 ('FRO', 2010),
 ('FRO', 2015),
 ('FSM', 2010),
 ('FSM', 2010),
 ('FSM', 2015),
 ('FSM', 2015),


## INTERSECT
<img src="images/intersevt_dia.png" style="width: 100px;"/>
The table form looks like this:
<img src="images/intersect_table.png" style="width: 150px;"/>
<code>INTERSECT</code> looks for records in common but not the individual key fields.

**Exercise**<br>
1. Determine all country codes and years that exist in both the economies or the populations tables. Order by code then year.

In [151]:
%%sql
select code, year
from economies
INTERSECT
select country_code, year
from populations
order by code,year

 * postgresql://postgres:***@localhost:5432/postgres
380 rows affected.


[('AFG', 2010),
 ('AFG', 2015),
 ('AGO', 2010),
 ('AGO', 2015),
 ('ALB', 2010),
 ('ALB', 2015),
 ('ARE', 2010),
 ('ARE', 2015),
 ('ARG', 2010),
 ('ARG', 2015),
 ('ARM', 2010),
 ('ARM', 2015),
 ('ATG', 2010),
 ('ATG', 2015),
 ('AUS', 2010),
 ('AUS', 2015),
 ('AUT', 2010),
 ('AUT', 2015),
 ('AZE', 2010),
 ('AZE', 2015),
 ('BDI', 2010),
 ('BDI', 2015),
 ('BEL', 2010),
 ('BEL', 2015),
 ('BEN', 2010),
 ('BEN', 2015),
 ('BFA', 2010),
 ('BFA', 2015),
 ('BGD', 2010),
 ('BGD', 2015),
 ('BGR', 2010),
 ('BGR', 2015),
 ('BHR', 2010),
 ('BHR', 2015),
 ('BHS', 2010),
 ('BHS', 2015),
 ('BIH', 2010),
 ('BIH', 2015),
 ('BLR', 2010),
 ('BLR', 2015),
 ('BLZ', 2010),
 ('BLZ', 2015),
 ('BOL', 2010),
 ('BOL', 2015),
 ('BRA', 2010),
 ('BRA', 2015),
 ('BRB', 2010),
 ('BRB', 2015),
 ('BRN', 2010),
 ('BRN', 2015),
 ('BTN', 2010),
 ('BTN', 2015),
 ('BWA', 2010),
 ('BWA', 2015),
 ('CAF', 2010),
 ('CAF', 2015),
 ('CAN', 2010),
 ('CAN', 2015),
 ('CHE', 2010),
 ('CHE', 2015),
 ('CHL', 2010),
 ('CHL', 2015),
 ('CHN', 2010),
 ('CHN', 2015),
 ('CIV', 2010),
 ('CIV', 2015),
 ('CMR', 2010),
 ('CMR', 2015),
 ('COD', 2010),
 ('COD', 2015),
 ('COG', 2010),
 ('COG', 2015),
 ('COL', 2010),
 ('COL', 2015),
 ('COM', 2010),
 ('COM', 2015),
 ('CPV', 2010),
 ('CPV', 2015),
 ('CRI', 2010),
 ('CRI', 2015),
 ('CYP', 2010),
 ('CYP', 2015),
 ('CZE', 2010),
 ('CZE', 2015),
 ('DEU', 2010),
 ('DEU', 2015),
 ('DJI', 2010),
 ('DJI', 2015),
 ('DMA', 2010),
 ('DMA', 2015),
 ('DNK', 2010),
 ('DNK', 2015),
 ('DOM', 2010),
 ('DOM', 2015),
 ('DZA', 2010),
 ('DZA', 2015),
 ('ECU', 2010),
 ('ECU', 2015),
 ('EGY', 2010),
 ('EGY', 2015),
 ('ERI', 2010),
 ('ERI', 2015),
 ('ESP', 2010),
 ('ESP', 2015),
 ('EST', 2010),
 ('EST', 2015),
 ('ETH', 2010),
 ('ETH', 2015),
 ('FIN', 2010),
 ('FIN', 2015),
 ('FJI', 2010),
 ('FJI', 2015),
 ('FRA', 2010),
 ('FRA', 2015),
 ('FSM', 2010),
 ('FSM', 2015),
 ('GAB', 2010),
 ('GAB', 2015),
 ('GBR', 2010),
 ('GBR', 2015),
 ('GEO', 2010),
 ('GEO', 2015),
 ('GHA', 2010),
 ('GHA', 2015),
 ('GIN', 2010),
 ('GIN', 2015),
 ('GMB', 2010),
 ('GMB', 2015),
 ('GNB', 2010),
 ('GNB', 2015),
 ('GNQ', 2010),
 ('GNQ', 2015),
 ('GRC', 2010),
 ('GRC', 2015),
 ('GRD', 2010),
 ('GRD', 2015),
 ('GTM', 2010),
 ('GTM', 2015),
 ('GUY', 2010),
 ('GUY', 2015),
 ('HKG', 2010),
 ('HKG', 2015),
 ('HND', 2010),
 ('HND', 2015),
 ('HRV', 2010),
 ('HRV', 2015),
 ('HTI', 2010),
 ('HTI', 2015),
 ('HUN', 2010),
 ('HUN', 2015),
 ('IDN', 2010),
 ('IDN', 2015),
 ('IND', 2010),
 ('IND', 2015),
 ('IRL', 2010),
 ('IRL', 2015),
 ('IRN', 2010),
 ('IRN', 2015),
 ('IRQ', 2010),
 ('IRQ', 2015),
 ('ISL', 2010),
 ('ISL', 2015),
 ('ISR', 2010),
 ('ISR', 2015),
 ('ITA', 2010),
 ('ITA', 2015),
 ('JAM', 2010),
 ('JAM', 2015),
 ('JOR', 2010),
 ('JOR', 2015),
 ('JPN', 2010),
 ('JPN', 2015),
 ('KAZ', 2010),
 ('KAZ', 2015),
 ('KEN', 2010),
 ('KEN', 2015),
 ('KGZ', 2010),
 ('KGZ', 2015),
 ('KHM', 2010),
 ('KHM', 2015),
 ('KIR', 2010),
 ('KIR', 2015),
 ('KNA', 2010),
 ('KNA', 2015),
 ('KOR', 2010),
 ('KOR', 2015),
 ('KWT', 2010),
 ('KWT', 2015),
 ('LAO', 2010),
 ('LAO', 2015),
 ('LBN', 2010),
 ('LBN', 2015),
 ('LBR', 2010),
 ('LBR', 2015),
 ('LBY', 2010),
 ('LBY', 2015),
 ('LCA', 2010),
 ('LCA', 2015),
 ('LKA', 2010),
 ('LKA', 2015),
 ('LSO', 2010),
 ('LSO', 2015),
 ('LTU', 2010),
 ('LTU', 2015),
 ('LUX', 2010),
 ('LUX', 2015),
 ('LVA', 2010),
 ('LVA', 2015),
 ('MAC', 2010),
 ('MAC', 2015),
 ('MAR', 2010),
 ('MAR', 2015),
 ('MDA', 2010),
 ('MDA', 2015),
 ('MDG', 2010),
 ('MDG', 2015),
 ('MDV', 2010),
 ('MDV', 2015),
 ('MEX', 2010),
 ('MEX', 2015),
 ('MHL', 2010),
 ('MHL', 2015),
 ('MKD', 2010),
 ('MKD', 2015),
 ('MLI', 2010),
 ('MLI', 2015),
 ('MLT', 2010),
 ('MLT', 2015),
 ('MMR', 2010),
 ('MMR', 2015),
 ('MNE', 2010),
 ('MNE', 2015),
 ('MNG', 2010),
 ('MNG', 2015),
 ('MOZ', 2010),
 ('MOZ', 2015),
 ('MRT', 2010),
 ('MRT', 2015),
 ('MUS', 2010),
 ('MUS', 2015),
 ('MWI', 2010),
 ('MWI', 2015),
 ('MYS', 2010),
 ('MYS', 2015),
 ('NAM', 2010),
 ('NAM', 2015),
 ('NER', 2010),
 ('NER', 2015),
 ('NGA', 2010),
 ('NGA', 2015),


**Exercise**<br>
As you think about major world cities and their corresponding country, you may ask which countries also have a city with the same name as their country name?

In [157]:
%%sql
select ci.name as city_and_country
from cities as ci
INTERSECT
select co.name
from countries as co.capital

 * postgresql://postgres:***@localhost:5432/postgres
2 rows affected.


[('Singapore',), ('Hong Kong',)]

## EXCEPT
<img src="images/except_dia.png" style="width: 100px;"/>
The table form looks like this:
<img src="images/except_table.png" style="width: 150px;"/>

**Exercise**<br>
* Get the names of cities in cities which are not noted as capital cities in countries as a single field result.

* Note that there are some countries in the world that are not included in the countries table, which will result in some cities not being labeled as capital cities when in fact they are.

In [160]:
%%sql
select name as not_capital_city
from cities
EXCEPT 
select capital
from countries
order by not_capital_city

 * postgresql://postgres:***@localhost:5432/postgres
170 rows affected.


[('Abidjan',),
 ('Ahmedabad',),
 ('Alexandria',),
 ('Almaty',),
 ('Auckland',),
 ('Bandung',),
 ('Barcelona',),
 ('Barranquilla',),
 ('Basra',),
 ('Belo Horizonte',),
 ('Bengaluru',),
 ('Bhopal',),
 ('Birmingham',),
 ('Brisbane',),
 ('Bucharest',),
 ('Busan',),
 ('Calgary',),
 ('Cali',),
 ('Caloocan',),
 ('Campinas',),
 ('Cape Town',),
 ('Casablanca',),
 ('Changchun',),
 ('Changsha',),
 ('Chaozhou',),
 ('Chengdu',),
 ('Chennai',),
 ('Chicago',),
 ('Chittagong',),
 ('Chongqing',),
 ('Cologne',),
 ('Cordoba',),
 ('Curitiba',),
 ('Daegu',),
 ('Daejeon',),
 ('Dalian',),
 ('Dallas',),
 ('Dar es Salaam',),
 ('Davao City',),
 ('Delhi',),
 ('Dongguan',),
 ('Douala',),
 ('Dubai',),
 ('Durban',),
 ('Ekurhuleni',),
 ('Faisalabad',),
 ('Fez',),
 ('Fortaleza',),
 ('Foshan',),
 ('Fukuoka',),
 ('Fuzhou',),
 ('Giza',),
 ('Guadalajara',),
 ('Guangzhou',),
 ('Guayaquil',),
 ('Gujranwala',),
 ('Hamburg',),
 ('Hangzhou',),
 ('Harbin',),
 ('Hefei',),
 ('Hiroshima',),
 ('Ho Chi Minh City',),
 ('Hong Kong',),
 ('Houston',),
 ('Hyderabad',),
 ('Hyderabad (India)',),
 ('Ibadan',),
 ('Incheon',),
 ('Isfahan',),
 ('Istanbul',),
 ('Izmir',),
 ('Jaipur',),
 ('Jeddah',),
 ('Jinan',),
 ('Johannesburg',),
 ('Kano',),
 ('Kanpur',),
 ('Kaohsiung',),
 ('Karachi',),
 ('Karaj',),
 ('Kawasaki',),
 ('Kharkov',),
 ('Kobe',),
 ('Kochi',),
 ('Kolkata',),
 ('Kwangju',),
 ('Kyoto',),
 ('Lagos',),
 ('Lahore',),
 ('Lanzhou',),
 ('Los Angeles',),
 ('Lucknow',),
 ('Makassar',),
 ('Mandalay',),
 ('Maracaibo',),
 ('Mashhad',),
 ('Medan',),
 ('Medellin',),
 ('Milan',),
 ('Monterrey',),
 ('Montreal',),
 ('Multan',),
 ('Mumbai',),
 ('Munich',),
 ('Nagoya',),
 ('Nagpur',),
 ('Nanjing',),
 ('New Taipei City',),
 ('New York City',),
 ('Ningbo',),
 ('Nizhny Novgorod',),
 ('Novosibirsk',),
 ('Omsk',),
 ('Oran',),
 ('Osaka',),
 ('Palembang',),
 ('Patna',),
 ('Peshawar',),
 ('Philadelphia',),
 ('Phoenix',),
 ('Porto Alegre',),
 ('Pune',),
 ('Qingdao',),
 ('Quanzhou',),
 ('Quezon City',),
 ('Rawalpindi',),
 ('Recife',),
 ('Rio de Janeiro',),
 ('Rosario',),
 ('Rostov-on-Don',),
 ('Saint Petersburg',),
 ('Saitama',),
 ('Salvador',),
 ('San Antonio',),
 ('San Diego',),
 ('Santa Cruz de la Sierra',),
 ('Sao Paulo',),
 ('Sapporo',),
 ('Semarang',),
 ('Shanghai',),
 ('Shantou',),
 ('Shenyang',),
 ('Shenzhen',),
 ('Shijiazhuang',),
 ('Shiraz',),
 ('Surabaya',),
 ('Surat',),
 ('Suzhou',),
 ('Tabriz',),
 ('Taichung',),
 ('Tainan',),
 ('Taipei',),
 ('Tangshan',),
 ("T'bilisi",),
 ('Tianjin',),
 ('Tijuana',),
 ('Toronto',),
 ('Ulsan',),
 ('Vijayawada',),
 ('Visakhapatnam',),
 ('Wenzhou',),
 ('Wuhan',),
 ('Xiamen',),
 ("Xi'an",),
 ('Yangon',),
 ('Yekaterinburg',),
 ('Yokohama',),
 ('Zhengzhou',),
 ('Zhongshan',),
 ('Zunyi',)]

**Exercise**<br>
* Determine the names of capital cities that are not listed in the cities table.

In [161]:
%%sql
select capital as not_in_cities
from countries
EXCEPT
select name
from cities
order by not_in_cities

 * postgresql://postgres:***@localhost:5432/postgres
136 rows affected.


[('Agana',),
 ('Amman',),
 ('Amsterdam',),
 ('Andorra la Vella',),
 ('Antananarivo',),
 ('Apia',),
 ('Ashgabat',),
 ('Asmara',),
 ('Astana',),
 ('Asuncion',),
 ('Athens',),
 ('Bamako',),
 ('Bandar Seri Begawan',),
 ('Bangui',),
 ('Banjul',),
 ('Basseterre',),
 ('Beirut',),
 ('Belmopan',),
 ('Bern',),
 ('Bishkek',),
 ('Bissau',),
 ('Bratislava',),
 ('Bridgetown',),
 ('Brussels',),
 ('Bujumbura',),
 ('Canberra',),
 ('Castries',),
 ('Charlotte Amalie',),
 ('Chisinau',),
 ('Colombo',),
 ('Conakry',),
 ('Copenhagen',),
 ('Damascus',),
 ('Djibouti',),
 ('Dodoma',),
 ('Doha',),
 ('Dublin',),
 ('Dushanbe',),
 ('Freetown',),
 ('Funafuti',),
 ('Gaborone',),
 ('Georgetown',),
 ('George Town',),
 ('Grand Turk',),
 ('Hamilton',),
 ('Helsinki',),
 ('Honiara',),
 ('Kathmandu',),
 ('Kigali',),
 ('Kingston',),
 ('Kingstown',),
 ('Koror',),
 ('Kuwait City',),
 ('La Paz',),
 ('Libreville',),
 ('Lilongwe',),
 ('Lisbon',),
 ('Ljubljana',),
 ('Lome',),
 ('Luxembourg',),
 ('Majuro',),
 ('Malabo',),
 ('Male',),
 ('Manama',),
 ('Maseru',),
 ('Mbabane',),
 ('Mogadishu',),
 ('Monaco',),
 ('Monrovia',),
 ('Moroni',),
 ('Muscat',),
 ('Nassau',),
 ('Naypyidaw',),
 ("N'Djamena",),
 ('New Delhi',),
 ('Niamey',),
 ('Nicosia',),
 ('Nouakchott',),
 ("Noum'ea",),
 ("Nuku'alofa",),
 ('Nuuk',),
 ('Oranjestad',),
 ('Oslo',),
 ('Ottawa',),
 ('Pago Pago',),
 ('Palikir',),
 ('Panama City',),
 ('Papeete',),
 ('Paramaribo',),
 ('Port-au-Prince',),
 ('Port Louis',),
 ('Port Moresby',),
 ('Port-of-Spain',),
 ('Porto-Novo',),
 ('Port-Vila',),
 ('Praia',),
 ('Pretoria',),
 ('Rabat',),
 ('Reykjavik',),
 ('Riga',),
 ('Road Town',),
 ('Roseau',),
 ("Saint George's",),
 ("Saint John's",),
 ('Saipan',),
 ('San Jose',),
 ('San Juan',),
 ('San Marino',),
 ('San Salvador',),
 ('Santo Domingo',),
 ('Sao Tome',),
 ('Sarajevo',),
 ('Skopje',),
 ('Sofia',),
 ('Stockholm',),
 ('Suva',),
 ('Tallinn',),
 ('Tarawa',),
 ('Tbilisi',),
 ('Tegucigalpa',),
 ('Thimphu',),
 ('Tirane',),
 ('Torshavn',),
 ('Ulaanbaatar',),
 ('Vaduz',),
 ('Valletta',),
 ('Victoria',),
 ('Vientiane',),
 ('Vilnius',),
 ('Washington D.C.',),
 ('Wellington',),
 ('Windhoek',),
 ('Yamoussoukro',),
 ('Yaren District',),
 ('Zagreb',),
 (None,)]

## Challenge Exercise

1. Identify the country codes that are included in either economies or currencies but not in populations.
2. Use that result to determine the names of cities in the countries that match the specification in the previous instruction.

In [7]:
%%sql
select code
from economies
UNION
select code
from currencies
EXCEPT
select country_code 
from populations

 * postgresql://postgres:***@localhost:5432/postgres
13 rows affected.


[('IOT',),
 ('SGS',),
 ('WLF',),
 ('SHN',),
 ('MSR',),
 ('CCK',),
 ('ROM',),
 ('NIU',),
 ('TWN',),
 ('TMP',),
 ('FLK',),
 ('AIA',),
 ('COK',)]

In [9]:
%%sql
select name, country_code
from cities
where country_code in (
    select code
    from economies
    UNION
    select code
    from currencies
    EXCEPT
    select country_code 
    from populations)

 * postgresql://postgres:***@localhost:5432/postgres
6 rows affected.


[('Bucharest', 'ROM'),
 ('Kaohsiung', 'TWN'),
 ('New Taipei City', 'TWN'),
 ('Taichung', 'TWN'),
 ('Tainan', 'TWN'),
 ('Taipei', 'TWN')]

# Nested queries
<code>Nested</code> or <code>sub</code> queries where queries are nested in queries.

These can be set up in: 
* <code>WHERE</code> clause.
* <code>SELECT</code> clause. <code>SELECT</code> sub query requires **aliasing** it.
* <code>FROM</code> clause.

## Inside WHERE 

**Exercise**
* Which countries had high average life expectancies (above 1.15) (at the country level) in 2015.

In [44]:
%%sql
--select country code and life_expectancy from populations
select co.name, po.country_code, year, life_expectancy
from populations as po
    inner join countries as co
        on po.country_code=co.code
where po.life_expectancy > 1.15 *
    (select avg(life_expectancy)
    from populations
    where year = 2015)
    and year=2015

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Australia', 'AUS', 2015, 82.4512),
 ('Switzerland', 'CHE', 2015, 83.1976),
 ('Spain', 'ESP', 2015, 83.3805),
 ('France', 'FRA', 2015, 82.6707),
 ('Hong Kong', 'HKG', 2015, 84.278),
 ('Iceland', 'ISL', 2015, 82.861),
 ('Italy', 'ITA', 2015, 83.4902),
 ('Japan', 'JPN', 2015, 83.8437),
 ('Singapore', 'SGP', 2015, 82.5951),
 ('Sweden', 'SWE', 2015, 82.5512)]

**Exercise**
* get the urban area population for only capital cities.

In [47]:
%%sql
select co.capital, co.name, ci.urbanarea_pop
from cities as ci
    inner join countries as co
        on co.code = ci.country_code
            where ci.name in (
                select capital
                from countries)
order by ci.urbanarea_pop desc

 * postgresql://postgres:***@localhost:5432/postgres
66 rows affected.


[('Beijing', 'China', 21516000.0),
 ('Dhaka', 'Bangladesh', 14543100.0),
 ('Tokyo', 'Japan', 13513700.0),
 ('Moscow', 'Russian Federation', 12197600.0),
 ('Cairo', 'Egypt', 10230400.0),
 ('Kinshasa', 'Congo, The Democratic Republic of the', 10130000.0),
 ('Jakarta', 'Indonesia', 10075300.0),
 ('Seoul', 'South Korea', 9995780.0),
 ('Mexico City', 'Mexico', 8974720.0),
 ('Lima', 'Peru', 8852000.0),
 ('London', 'United Kingdom', 8673710.0),
 ('Bangkok', 'Thailand', 8280920.0),
 ('Tehran', 'Iran', 8154050.0),
 ('Bogota', 'Colombia', 7878780.0),
 ('Baghdad', 'Iraq', 7180890.0),
 ('Hanoi', 'Vietnam', 6844100.0),
 ('Santiago', 'Chile', 5743720.0),
 ('Riyadh', 'Saudi Arabia', 5676620.0),
 ('Singapore', 'Singapore', 5535000.0),
 ('Ankara', 'Turkey', 5271000.0),
 ('Khartoum', 'Sudan', 3639600.0),
 ('Berlin', 'Germany', 3517420.0),
 ('Algiers', 'Algeria', 3415810.0),
 ('Kabul', 'Afghanistan', 3414100.0),
 ('Pyongyang', 'North Korea', 3255390.0),
 ('Madrid', 'Spain', 3207250.0),
 ('Baku', 'Azerbaijan', 3202300.0),
 ('Nairobi', 'Kenya', 3138370.0),
 ('Addis Ababa', 'Ethiopia', 3103670.0),
 ('Buenos Aires', 'Argentina', 3054300.0),
 ('Kiev', 'Ukraine', 2908700.0),
 ('Rome', 'Italy', 2877220.0),
 ('Luanda', 'Angola', 2825310.0),
 ('Quito', 'Ecuador', 2671190.0),
 ('Managua', 'Nicaragua', 2560790.0),
 ('Brasilia', 'Brazil', 2556150.0),
 ('Yaounde', 'Cameroon', 2440460.0),
 ('Tashkent', 'Uzbekistan', 2309600.0),
 ('Phnom Penh', 'Cambodia', 2234570.0),
 ('Paris', 'France', 2229620.0),
 ('Ouagadougou', 'Burkina Faso', 2200000.0),
 ('Guatemala City', 'Guatemala', 2110100.0),
 ('Havana', 'Cuba', 2106150.0),
 ('Accra', 'Ghana', 2070460.0),
 ('Minsk', 'Belarus', 1959780.0),
 ('Caracas', 'Venezuela', 1943900.0),
 ("Sana'a", 'Yemen', 1937450.0),
 ('Islamabad', 'Pakistan', 1900000.0),
 ('Vienna', 'Austria', 1863880.0),
 ('Brazzaville', 'Congo', 1827000.0),
 ('Manila', 'Philippines', 1780150.0),
 ('Kuala Lumpur', 'Malaysia', 1768000.0),
 ('Maputo', 'Mozambique', 1766180.0),
 ('Budapest', 'Hungary', 1759410.0),
 ('Warsaw', 'Poland', 1753980.0),
 ('Lusaka', 'Zambia', 1742980.0),
 ('Harare', 'Zimbabwe', 1606000.0),
 ('Kampala', 'Uganda', 1507080.0),
 ('Prague', 'Czech Republic', 1324000.0),
 ('Montevideo', 'Uruguay', 1305080.0),
 ('Abuja', 'Nigeria', 1235880.0),
 ('Dakar', 'Senegal', 1146050.0),
 ('Abu Dhabi', 'United Arab Emirates', 1145000.0),
 ('Tripoli', 'Libya', 1126000.0),
 ('Yerevan', 'Armenia', 1060140.0),
 ('Tunis', 'Tunisia', 1056250.0)]

## Inside SELECT
Nested SELECT clauses must be **ALIASED**

**Exercise**
* select the top nine countries in terms of number of cities appearing in the cities table.

In [66]:
%%sql
--solving it via inner join
select co.name as country, count(*) as n_cities
from cities as ci
inner join countries as co
on co.code = ci.country_code
group by country
order by n_cities desc, country
limit 9

 * postgresql://postgres:***@localhost:5432/postgres
9 rows affected.


[('China', 36),
 ('India', 18),
 ('Japan', 11),
 ('Brazil', 10),
 ('Pakistan', 9),
 ('United States', 9),
 ('Indonesia', 7),
 ('Russian Federation', 7),
 ('South Korea', 7)]

In [74]:
%%sql
--solving it via nested select
select co.name as country, (
    select count(*)
    from cities as ci
    where co.code = ci.country_code) as n_cities
from countries as co
order by n_cities desc, country
limit 9

 * postgresql://postgres:***@localhost:5432/postgres
9 rows affected.


[('China', 36),
 ('India', 18),
 ('Japan', 11),
 ('Brazil', 10),
 ('Pakistan', 9),
 ('United States', 9),
 ('Indonesia', 7),
 ('Russian Federation', 7),
 ('South Korea', 7)]

## Inside FROM

**Exercise**
* determine the number of languages spoken for each country, identified by the country's local name!

In [79]:
%%sql
select co.local_name, subquery.n_languages
from countries as co, (
    select code, count(*) as n_languages
    from languages
    group by code) as subquery 
where co.code = subquery.code
order by n_languages desc

 * postgresql://postgres:***@localhost:5432/postgres
199 rows affected.


[('Zambia', 19),
 ('YeItyop´iya', 16),
 ('Zimbabwe', 16),
 ('Bharat/India', 14),
 ('Nepal', 14),
 ('Mali', 13),
 ('France', 13),
 ('South Africa', 13),
 ('Malawi', 12),
 ('Angola', 12),
 ('Pakistan', 11),
 ('Panama', 11),
 ('Ghana', 11),
 ('Singapore/Singapura/Xinjiapo/Singapur', 10),
 ('Kypros/Kibris', 10),
 ('Canada', 10),
 ('Schweiz/Suisse/Svizzera/Svizra', 10),
 ('Belize', 9),
 ('Australia', 9),
 ('Zhongquo', 9),
 ('Micronesia', 9),
 ('Namibia', 9),
 ('Botswana', 9),
 ('New Zealand/Aotearoa', 9),
 ('Mocambique', 8),
 ('Sao Tome e Principe', 8),
 ('Iran', 8),
 ('Malaysia', 8),
 ('Suriya', 7),
 ('Aruba', 7),
 ('Bolivia', 7),
 ('Moldova', 7),
 ('Magyarorszag', 7),
 ('Al-Maghrib', 7),
 ('Northern Mariana Islands', 7),
 ('Senegal/Sounougal', 7),
 ('Ertra', 7),
 ('Nigeria', 6),
 ('Uganda', 6),
 ('Makedonija', 6),
 ('Belau/Palau', 6),
 ('Le Gabon', 6),
 ('Peru/Piruw', 6),
 ('Al-Jaza\x92ir/Algerie', 6),
 ('Argentina', 6),
 ('Guam', 6),
 ('Rwanda/Urwanda', 6),
 ('Burundi/Uburundi', 6),
 ('Suriname', 5),
 ('Vanuatu', 5),
 ('´Uman', 5),
 ('Bosna i Hercegovina', 5),
 ('Amerika Samoa', 5),
 ('Balgarija', 5),
 ('Cayman Islands', 5),
 ('Chile', 5),
 ('Trinidad and Tobago', 5),
 ('Espana', 5),
 ('The Gambia', 5),
 ('Sakartvelo', 5),
 ('Togo', 5),
 ('Eesti', 5),
 ('Guyana', 5),
 ('Xianggang/Hong Kong', 5),
 ('Al-´Iraq', 5),
 ('Al-Imarat al-´Arabiya al-Muttahida', 5),
 ('Osterreich', 5),
 ('Mauritius', 5),
 ('As-Sudan', 5),
 ('Slovenija', 5),
 ('Slovensko', 5),
 ('Republique Democratique du Congo', 5),
 ('Sesel/Seychelles', 5),
 ('Lietuva', 5),
 ('Luxembourg/Letzebuerg', 5),
 ('Muritaniya/Mauritanie', 5),
 ('Guine-Bissau', 4),
 ('Danmark', 4),
 ('Shqiperia', 4),
 ('Afganistan/Afqanestan', 4),
 ('Gibraltar', 4),
 ('Andorra', 4),
 ('Kyrgyzstan', 4),
 ('Tonga', 4),
 ('Sierra Leone', 4),
 ('Ecuador', 4),
 ('Azarbaycan', 4),
 ('Belarus', 4),
 ('Malta', 4),
 ('Tchad/Tshad', 4),
 ('Turkmenostan', 4),
 ('Hrvatska', 4),
 ('Djibouti/Jibuti', 4),
 ('Tuvalu', 4),
 ('Druk-Yul', 4),
 ('Lao', 4),
 ('Latvija', 4),
 ('Lesotho', 4),
 ('Lubnan', 4),
 ('Uzbekiston', 4),
 ('Libiya', 4),
 ('Rossija', 4),
 ('Monaco', 4),
 ('Indonesia', 4),
 ('Tanzania', 4),
 ('Island', 4),
 ('Polska', 4),
 ('Italia', 4),
 ('Al-Bahrayn', 4),
 ('Papua New Guinea/Papua Niugini', 4),
 ('Suomi', 4),
 ('Nicaragua', 4),
 ('Soomaaliya', 4),
 ('Guinea Ecuatorial', 4),
 ('Prathet Thai', 3),
 ('Hajastan', 3),
 ('Belgie/Belgique', 3),
 ('Benin', 3),
 ('Brunei Darussalam', 3),
 ('Pilipinas', 3),
 ('Foroyar', 3),
 ('Kalaallit Nunaat/Gronland', 3),
 ('Yisra\x92el/Isra\x92il', 3),
 ('Cameroun/Cameroon', 3),
 ('Kenya', 3),
 ('Centrafrique/Be-Afrika', 3),
 ('Komori/Comores', 3),
 ('Congo', 3),
 ('Liechtenstein', 3),
 ('Madagasikara/Madagascar', 3),
 ('Mexico', 3),
 ('Mongol Uls', 3),
 ('Naoero/Nauru', 3),
 ('Niger', 3),
 ('Norge', 3),
 ('Polynesie francaise', 3),
 ('Solomon Islands', 3),
 ('Sri Lanka/Ilankai', 3),
 ('Tunis/Tunisie', 3),
 ('Turkiye', 3),
 ('Ukrajina', 3),
 ('Uruguay', 3),
 ('Viet Nam', 3),
 ('United States', 3),
 ('Taehan Min\x92guk (Namhan)', 2),
 ('Costa Rica', 2),
 ('Burkina Faso', 2),
 ('Barbados', 2),
 ('Saint Lucia', 2),
 ('Antigua and Barbuda', 2),
 ('Nouvelle-Caledonie', 2),
 ('Kiribati', 2),
 ('Paraguay', 2),
 ('Cote d\x92Ivoire', 2),
 ('Bermuda', 2),
 ('Portugal', 2),
 ('Puerto Rico', 2),
 ('Marshall Islands/Majol', 2),
 ('Dhivehi Raajje/Maldives', 2),
 ('Qatar', 2),
 ('Liberia', 2),
 ('Qazaqstan', 2),
 ('Kampuchea', 2),
 ('Al-Urdunn', 2),
 ('Bangladesh', 2),
 ('Samoa', 2),
 ('Saint Vincent and the Grenadines', 2),
 ('Jamaica', 2),
 ('kaNgwane', 2),
 ('Ireland/Eire', 2),
 ('Honduras', 2),
 ('Tocikiston', 2),
 ('Haiti/Dayti', 2),
 ('Ellada', 2),
 ('Guatemala', 2),
 ('Grenada', 2),
 ('El Salvador', 2),
 ('Misr', 2),
 ('Al-Kuwayt', 2),
 ('Venezuela', 2),
 ('The Bahamas', 2),
 ('Dominica', 2),
 ('Myanma Pye', 1),
 ('Sverige', 1),
 ('Saint Kitts and Nevis', 1),
 ('Deutschland', 1),
 ('Cuba', 1),
 ('San Marino', 1),
 ('Al-´Arabiya as-Sa´udiya', 1),
 ('Choson Minjujuui In´min Ko

## Multiple sub-queries
**Exercise**
* for each of the six continents listed in 2015, you'll identify which country had the maximum inflation rate (and how high it was) using multiple subqueries. 

In [88]:
%%sql
select name, continent, inflation_rate
from countries
    inner join economies
        on countries.code = economies.code
where year = 2015 and inflation_rate in (
    select max(inflation_rate) as max_inf
    from (
        select co.name, continent, ec.inflation_rate
        from countries as co
            inner join economies as ec
                using(code)
                    where year = 2015) as subquery
    group by continent)

 * postgresql://postgres:***@localhost:5432/postgres
6 rows affected.


[('Haiti', 'North America', 7.524),
 ('Malawi', 'Africa', 21.858),
 ('Nauru', 'Oceania', 9.784),
 ('Ukraine', 'Europe', 48.684),
 ('Venezuela', 'South America', 121.738),
 ('Yemen', 'Asia', 39.403)]

**Challenge**
* get 2015 economic data for countries that do **not** have 
    1. <code>gov_form</code> of 'Constitutional Monarchy' or word 'Republic' in it

In [97]:
%%sql
select code, inflation_rate, unemployment_rate
from economies
where year = 2015 and code not in (
    select code
    from countries
    where (gov_form = 'Constitutional Monarchy' or gov_form like '%Republic%'))
order by inflation_rate

 * postgresql://postgres:***@localhost:5432/postgres
26 rows affected.


[('AFG', -1.549, None),
 ('CHE', -1.14, 3.178),
 ('PRI', -0.751, 12.0),
 ('ROU', -0.596, 6.812),
 ('BRN', -0.423, 6.9),
 ('TON', -0.283, None),
 ('OMN', 0.065, None),
 ('TLS', 0.553, None),
 ('BEL', 0.62, 8.492),
 ('CAN', 1.132, 6.9),
 ('MNE', 1.204, None),
 ('SRB', 1.392, 18.2),
 ('AUS', 1.461, 6.058),
 ('QAT', 1.814, None),
 ('BHR', 1.836, None),
 ('WSM', 1.923, None),
 ('MYS', 2.104, 3.1),
 ('SAU', 2.189, 5.591),
 ('HKG', 3.037, 3.296),
 ('KWT', 3.233, 2.072),
 ('ARE', 4.07, None),
 ('MAC', 4.564, 1.825),
 ('SWZ', 4.96, None),
 ('BTN', 6.336, 3.2),
 ('LBY', 9.839, None),
 ('SSD', 52.813, None)]

**Challenge**
* get the country names and other 2015 data in the economies table and the countries table for Central American countries with an official language.

In [37]:
%%sql
select co.name, gdp_percapita, inflation_rate
from countries as co
    left join economies as ec
        on co.code = ec.code and ec.code in (
            select code
            from languages
            where official = 'True')
where ec.year=2015 and co.region = 'Central America'

 * postgresql://postgres:***@localhost:5432/postgres
7 rows affected.


[('Nicaragua', 2086.89, 3.997),
 ('Costa Rica', 11436.0, 0.802),
 ('Belize', 4757.11, -0.862),
 ('Guatemala', 3921.87, 2.389),
 ('Panama', 13113.7, 0.134),
 ('El Salvador', 4217.0, -0.731),
 ('Honduras', 2567.05, 3.158)]

**Challenge**

* calculate the average fertility rate for each region in 2015.

In [52]:
%%sql
select region, avg(fertility_rate) as avg_fertility
from countries as co
    inner join populations as po
        on co.code = po.country_code
where year = 2015
group by region
order by avg_fertility

 * postgresql://postgres:***@localhost:5432/postgres
23 rows affected.


[('Southern Europe', 1.42610000371933),
 ('Eastern Europe', 1.49088890022702),
 ('Baltic Countries', 1.60333331425985),
 ('Eastern Asia', 1.62071430683136),
 ('Western Europe', 1.6325000077486),
 ('North America', 1.76575002074242),
 ('British Islands', 1.875),
 ('Nordic Countries', 1.89333335558573),
 ('Australia and New Zealand', 1.91149997711182),
 ('Caribbean', 1.95057143483843),
 ('Southeast Asia', 2.15600001811981),
 ('South America', 2.27475001414617),
 ('Central America', 2.32637499272823),
 ('Middle East', 2.54705556895998),
 ('Southern and Central Asia', 2.63414285864149),
 ('Micronesia', 2.86475002765656),
 ('Northern Africa', 2.9081666469574),
 ('Southern Africa', 2.99079999923706),
 ('Melanesia', 3.13579998016357),
 ('Polynesia', 3.24433326721191),
 ('Eastern Africa', 4.38670586838442),
 ('Western Africa', 4.96012498438358),
 ('Central Africa', 4.96788883209229)]

**Challenge**
*  determine the top 10 capital cities in Europe and the Americas in terms of a calculated percentage using <code>city_proper_pop</code> and <code>metroarea_pop</code> in <code>cities</code>

In [60]:
%%sql
select ci.name, city_proper_pop / metroarea_pop * 100 as city_pct_metroarea
from cities as ci
    inner join countries as co
        on ci.country_code = co.code and ci.name in (
            select capital
            from countries
            where continent = 'Europe' or continent like '%America')
where city_proper_pop is not null and metroarea_pop is not null
order by city_pct_metroarea desc
limit 10

 * postgresql://postgres:***@localhost:5432/postgres
10 rows affected.


[('Lima', 82.3441863059998),
 ('Bogota', 80.3957462310791),
 ('Moscow', 75.4334926605225),
 ('Vienna', 71.6877281665802),
 ('Montevideo', 67.0096158981323),
 ('Caracas', 66.4818167686462),
 ('Rome', 66.0855233669281),
 ('Brasilia', 65.2101457118988),
 ('London', 62.4918222427368),
 ('Budapest', 60.090184211731)]